{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial: Identifying low-frequency somatic mutations in FGFR2 with UMI-tagged smMIPs \n",
    "\n",
    "One of the main use cases for amplimap is to accurately quantify the level of mutations found at very low allele frequencies. This is a common challenge when working with somatic mutations, for example in tumours from cancer samples.\n",
    "\n",
    "For this tutorial, we will look at data from an experiment described in [Selfish mutations dysregulating RAS-MAPK signaling are pervasive in aged human testes](https://www.biorxiv.org/content/early/2018/05/04/314815).\n",
    "Maher et al conducted a screen for somatic mutations in testes, searching for mutations observed at very low frequencies (<3%).\n",
    "Due to the technical challenges inherent in calling mutations at such low frequencies, possible hits needed to be validated with a separate method.\n",
    "\n",
    "To perform this validation, Maher et al resequenced regions around observed mutations using single-molecule molecular inversion probes (smMIPs). Each smMIP was tagged with unique molecular identifiers (UMIs), enabling accurate quantification of variant allele frequencies (VAFs).\n",
    "\n",
    "Here, we will look at how amplimap can process data from such an experiment and generate VAFs for specific mutations.\n",
    "While the biological system being studied here is slightly different from somatic mutations in cancer, they can be processed and analysed in exactly the same way.\n",
    "\n",
    "## Analysis overview\n",
    "\n",
    "Starting from the raw sequencing reads, we would like to:\n",
    "\n",
    "- Trim off UMIs and extension/ligation arm sequences\n",
    "- Align reads to the reference genome\n",
    "- Group reads with the same UMI together and determine a consensus call for each nucleotide\n",
    "- Pile up reads in the target regions and count how often we observed each nucleotide at each position\n",
    "- Compare the observed nucleotide counts to our expectation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initial setup\n",
    "\n",
    "To run this tutorial amplimap needs to be installed and configured already.\n",
    "Please see [Installation](https://amplimap.readthedocs.io/en/latest/installation.html)\n",
    "and [Configuration](https://amplimap.readthedocs.io/en/latest/configuration.html) for details.\n",
    "\n",
    "In particular, you need to have the hg19 (GRCh37) reference genome and the\n",
    "associated indices prepared for use with your default aligner (see [Reference genome paths](https://amplimap.readthedocs.io/en/latest/configuration.html#reference-genome-paths))."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparing the working directory\n",
    "\n",
    "For every experiment that we want to process, we create a new working\n",
    "directory. This will contain all the input files required, as well as\n",
    "the output generated by amplimap. This makes it easy to keep track of\n",
    "the data for each experiment, as well as to rerun analyses if required.\n",
    "\n",
    "To create a directory, we use the standard ``mkdir`` unix command and\n",
    "change into it with ``cd``:\n",
    "\n",
    "    mkdir FGFR2\n",
    "    cd FGFR2\n",
    "\n",
    "All further commands should now be run inside this working directory.\n",
    "\n",
    "### reads_in\n",
    "\n",
    "The first input we need to provide to amplimap is of course the\n",
    "sequencing data. These can be obtained directly from the sequencer as\n",
    "``.fastq.gz`` files and should be placed in a directory called [``reads_in``](https://amplimap.readthedocs.io/en/latest/usage.html#reads-in).\n",
    "\n",
    "[Download the sample data from this tutorial](http://userweb.molbiol.ox.ac.uk/public/koelling/amplimap/tutorial_data/FGFR2.tar)\n",
    "and extract the ``reads_in`` directory into your working directory. There are many different ways of doing this but\n",
    "we recommend using ``wget`` and ``tar`` on the command line:\n",
    "\n",
    "    wget http://userweb.molbiol.ox.ac.uk/public/koelling/amplimap/tutorial_data/FGFR2.tar\n",
    "    tar xf FGFR2.tar\n",
    "\n",
    "You can use ``ls`` to check that the files have been extracted to the correct subdirectory:\n",
    "\n",
    "    ls reads_in\n",
    "\n",
    "This should display a list of ten fastq.gz files, which represent read 1\n",
    "and read 2 of five samples:\n",
    "\n",
    "    1D29_L001_R1_001.fastq.gz\n",
    "    1D29_L001_R2_001.fastq.gz\n",
    "    1D7_L001_R1_001.fastq.gz\n",
    "    1D7_L001_R2_001.fastq.gz\n",
    "    4C22_L001_R1_001.fastq.gz\n",
    "    4C22_L001_R2_001.fastq.gz\n",
    "    4C23_L001_R1_001.fastq.gz\n",
    "    4C23_L001_R2_001.fastq.gz\n",
    "    979-blood_L001_R1_001.fastq.gz\n",
    "    979-blood_L001_R2_001.fastq.gz\n",
    "    \n",
    "### probes_mipgen.csv\n",
    "\n",
    "Next, we need to provide a [probes.csv file](https://amplimap.readthedocs.io/en/latest/usage.html#probes-csv) that describes the used\n",
    "primer sequences and the regions they are supposed to capture.\n",
    "In this case, the smMIPs were designed using [MIPGEN](http://shendurelab.github.io/MIPGEN/), which\n",
    "already generated such a table, albeit in a slightly different format.\n",
    "\n",
    "amplimap can automatically convert a MIPGEN probe table saved in CSV format into\n",
    "the amplimap format. \n",
    "\n",
    "Create a new plain text file called ``probes_mipgen.csv`` (for example using ``nano``\n",
    "or ``vim``) in your working directory and copy the following text into it:\n",
    "\n",
    "    chr,ext_probe_sequence,lig_probe_sequence,mip_scan_start_position,mip_scan_stop_position,probe_strand,mip_name\n",
    "    chr10,GCAGTCAACCAAGAAAAG,CCTCAAAAGTTACATTCCGAAT,123276799,123276910,+,2_54_FGFR2_p.S236C\n",
    "    chr10,CTTGTTTTCTAGGCCGCC,ACTCTGCATGGTTGACAGTTCT,123276860,123276971,-,2_55_FGFR2_p.C227S\n",
    "    \n",
    "This is a copy of a probe table created by MIPGEN, including two probes and a subset of the available columns.\n",
    "As long as these columns are provided, it will be converted into the correct format when we run amplimap.\n",
    "Any read pair that does not match the primers (arm sequences) specified for either of these probes\n",
    "will be ignored.\n",
    "    \n",
    "### targets.csv\n",
    "\n",
    "We also need a [file describing the target regions](https://amplimap.readthedocs.io/en/latest/usage.html#targets-csv)\n",
    "that amplimap should analyse. amplimap will generate pileups for all regions listed in this file,\n",
    "and ignore all reads outside.\n",
    "We can provide this file either in BED format (with no column names and 0-based coordinates)\n",
    "or in CSV format (with column names and 1-based coordinates). In this case we\n",
    "are doing the latter.\n",
    "\n",
    "Create a new plain text file called ``targets.csv`` (for example using ``nano``\n",
    "or ``vim``) in your working directory and copy the following text into it:\n",
    "\n",
    "    chr,start,end,id\n",
    "    chr10,123276799,123276971,FGFR2_C342A\n",
    "\n",
    "This specifies a single target region, spanning from chr10:123276799-123276971 and called ``FGFR2_C342A``.\n",
    "\n",
    "### config.yaml\n",
    "\n",
    "Finally, we create a config.yaml file to set some experiment-specific settings.\n",
    "We could set [a lot more options](https://amplimap.readthedocs.io/en/latest/configuration.html)\n",
    "here but in this case set a few of them. All the other options will be left at\n",
    "the as specified in the default configuration.\n",
    "\n",
    "Create a new plain text file called ``config.yaml`` (for example using ``nano``\n",
    "or ``vim``) in your working directory and copy the following text into it:\n",
    "\n",
    "    general:\n",
    "      genome_name: \"hg19\"\n",
    "      umi_min_consensus_count: 2\n",
    "\n",
    "    parse_reads:\n",
    "      umi_one: 5\n",
    "      umi_two: 5\n",
    "\n",
    "The first setting tells amplimap to run using the reference genome hg19, as specified in your\n",
    "[default configuration](https://amplimap.readthedocs.io/en/latest/configuration.html#default-configuration).\n",
    "If you do not have this reference genome set up there, you can also specify the necessary paths directly\n",
    "in the ``config.yaml``, by adding the following additional lines and editing the paths to match your local setup:\n",
    "\n",
    "    paths:\n",
    "      hg19:\n",
    "        bwa: \"/INSERT/PATH/TO/PREFIX\"\n",
    "        fasta: \"/INSERT/PATH/TO/FASTA\"\n",
    "        \n",
    "For ``bwa`` you would provide the path of the prefix you used when [building the BWA index](http://bio-bwa.sourceforge.net/bwa.shtml#3).\n",
    "If you are using Bowtie2 as your aligner you can replace ``bwa`` with ``bowtie2`` in the config file and provide the prefix of its index files instead.\n",
    "For ``fasta`` you would provide the path to the corresponding FASTA file, which needs to have been indexed with ``samtools faidx``.\n",
    "\n",
    "The ``umi_min_consensus_count`` specifies how many concordant reads are required for a UMI group to be counted.\n",
    "By setting this to two we will ignore all UMI groups that are only supported by one read, or where we don't have\n",
    "at least two reads that agree.\n",
    "\n",
    "Finally, the ``umi_one`` and ``umi_two`` settings under ``parse_reads`` tell amplimap how long the UMI sequences on read one\n",
    "and read two are. It is important to get this setting right, since setting an incorrect value here will probably result in none of the reads being matched to the expected probe arms. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running amplimap\n",
    "\n",
    "Now we can run amplimap using the ``pileups`` target. First we will do a dry-run to confirm that all input files can be found:\n",
    "\n",
    "    amplimap pileups\n",
    "    \n",
    "This should output a long list of commands, ending with these lines:\n",
    "\n",
    "    Job counts:\n",
    "        count\tjobs\n",
    "        5\talign_pe\n",
    "        1\tconvert_mipgen_probes\n",
    "        1\tconvert_targets_csv\n",
    "        5\tdo_pileup\n",
    "        10\tlink_reads\n",
    "        5\tparse_reads_pe\n",
    "        1\tpileup_agg\n",
    "        1\tpileups\n",
    "        1\tstart_analysis\n",
    "        5\tstats_alignment\n",
    "        1\tstats_alignment_agg\n",
    "        1\tstats_reads_agg\n",
    "        1\tstats_samples_agg\n",
    "        2\ttool_version\n",
    "        40        \n",
    "    amplimap dry run successful. Set --run to run!\n",
    "\n",
    "You can see how amplimap is planning to run 5 alignment jobs (align_pe) and 5 pileups (do_pileup), corresponding to the 5 samples we are analysing.\n",
    "\n",
    "Having confirmed that everything looks as expected, we can run amplimap:\n",
    "\n",
    "    amplimap pileups --run\n",
    "\n",
    "This will take a few minutes to complete. It would be much faster if we\n",
    "ran jobs in parallel (for example using a cluster), but we are not\n",
    "doing that for the purposes of this tutorial."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysing the results\n",
    "\n",
    "amplimap has now processed our reads, aligned them to the reference genome and calculated read counts for each site and each sample.\n",
    "All of the output files have been placed into the ``analysis`` directory.\n",
    "\n",
    "Let's explore some of the output. Most analyses in amplimap produce one or more CSV file with a table of results. In this tutorial, we will use Python and pandas to process and visualize these files. However, the same thing could also be done in R or Excel."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### analysis/reads_parsed/\n",
    "\n",
    "As a first step, let's have a look at the ``stats_reads.csv`` file, which tells us some general statistics about the number of reads and the number of UMI groups for the different probes and samples. It contains one row for each sample/probe combination, with a set of columns providing details such as the raw number of read pairs (``read_pairs``), the number of unique UMIs (``umis_total``), the min/mean/max number of read pairs supporting a UMI group (``umis_coverage_min/mean/max``), the different percentiles of the UMI coverage and the number of UMIs that had at least 2/3/5/10 supporting read pairs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>probe</th>\n",
       "      <th>read_pairs</th>\n",
       "      <th>umis_total</th>\n",
       "      <th>umis_coverage_min</th>\n",
       "      <th>umis_coverage_mean</th>\n",
       "      <th>umis_coverage_max</th>\n",
       "      <th>umis_coverage_05pct</th>\n",
       "      <th>umis_coverage_25pct</th>\n",
       "      <th>umis_coverage_median</th>\n",
       "      <th>umis_coverage_75pct</th>\n",
       "      <th>umis_coverage_95pct</th>\n",
       "      <th>umis_coverage_ge_2</th>\n",
       "      <th>umis_coverage_ge_3</th>\n",
       "      <th>umis_coverage_ge_5</th>\n",
       "      <th>umis_coverage_ge_10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1D29</td>\n",
       "      <td>2_54_FGFR2_p.S236C</td>\n",
       "      <td>25540</td>\n",
       "      <td>7829</td>\n",
       "      <td>1</td>\n",
       "      <td>3.262230</td>\n",
       "      <td>34</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4211</td>\n",
       "      <td>2740</td>\n",
       "      <td>1084</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1D29</td>\n",
       "      <td>2_55_FGFR2_p.C227S</td>\n",
       "      <td>41205</td>\n",
       "      <td>11506</td>\n",
       "      <td>1</td>\n",
       "      <td>3.581175</td>\n",
       "      <td>43</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6748</td>\n",
       "      <td>4590</td>\n",
       "      <td>1941</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1D7</td>\n",
       "      <td>2_54_FGFR2_p.S236C</td>\n",
       "      <td>22888</td>\n",
       "      <td>8308</td>\n",
       "      <td>1</td>\n",
       "      <td>2.754935</td>\n",
       "      <td>39</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3566</td>\n",
       "      <td>2091</td>\n",
       "      <td>692</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1D7</td>\n",
       "      <td>2_55_FGFR2_p.C227S</td>\n",
       "      <td>34098</td>\n",
       "      <td>11237</td>\n",
       "      <td>1</td>\n",
       "      <td>3.034440</td>\n",
       "      <td>88</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5395</td>\n",
       "      <td>3303</td>\n",
       "      <td>1265</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4C22</td>\n",
       "      <td>2_54_FGFR2_p.S236C</td>\n",
       "      <td>22911</td>\n",
       "      <td>8689</td>\n",
       "      <td>1</td>\n",
       "      <td>2.636782</td>\n",
       "      <td>80</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3485</td>\n",
       "      <td>1887</td>\n",
       "      <td>619</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4C22</td>\n",
       "      <td>2_55_FGFR2_p.C227S</td>\n",
       "      <td>31577</td>\n",
       "      <td>11080</td>\n",
       "      <td>1</td>\n",
       "      <td>2.849910</td>\n",
       "      <td>44</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4897</td>\n",
       "      <td>2830</td>\n",
       "      <td>972</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>4C23</td>\n",
       "      <td>2_54_FGFR2_p.S236C</td>\n",
       "      <td>24682</td>\n",
       "      <td>8136</td>\n",
       "      <td>1</td>\n",
       "      <td>3.033677</td>\n",
       "      <td>46</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4027</td>\n",
       "      <td>2478</td>\n",
       "      <td>911</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>4C23</td>\n",
       "      <td>2_55_FGFR2_p.C227S</td>\n",
       "      <td>33085</td>\n",
       "      <td>10528</td>\n",
       "      <td>1</td>\n",
       "      <td>3.142572</td>\n",
       "      <td>40</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5421</td>\n",
       "      <td>3367</td>\n",
       "      <td>1247</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>979-blood</td>\n",
       "      <td>2_54_FGFR2_p.S236C</td>\n",
       "      <td>22408</td>\n",
       "      <td>8814</td>\n",
       "      <td>1</td>\n",
       "      <td>2.542319</td>\n",
       "      <td>122</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3313</td>\n",
       "      <td>1828</td>\n",
       "      <td>563</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>979-blood</td>\n",
       "      <td>2_55_FGFR2_p.C227S</td>\n",
       "      <td>37032</td>\n",
       "      <td>12668</td>\n",
       "      <td>1</td>\n",
       "      <td>2.923271</td>\n",
       "      <td>142</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5631</td>\n",
       "      <td>3401</td>\n",
       "      <td>1252</td>\n",
       "      <td>159</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample               probe  read_pairs  umis_total  umis_coverage_min  \\\n",
       "0       1D29  2_54_FGFR2_p.S236C       25540        7829                  1   \n",
       "1       1D29  2_55_FGFR2_p.C227S       41205       11506                  1   \n",
       "2        1D7  2_54_FGFR2_p.S236C       22888        8308                  1   \n",
       "3        1D7  2_55_FGFR2_p.C227S       34098       11237                  1   \n",
       "4       4C22  2_54_FGFR2_p.S236C       22911        8689                  1   \n",
       "5       4C22  2_55_FGFR2_p.C227S       31577       11080                  1   \n",
       "6       4C23  2_54_FGFR2_p.S236C       24682        8136                  1   \n",
       "7       4C23  2_55_FGFR2_p.C227S       33085       10528                  1   \n",
       "8  979-blood  2_54_FGFR2_p.S236C       22408        8814                  1   \n",
       "9  979-blood  2_55_FGFR2_p.C227S       37032       12668                  1   \n",
       "\n",
       "   umis_coverage_mean  umis_coverage_max  umis_coverage_05pct  \\\n",
       "0            3.262230                 34                  1.0   \n",
       "1            3.581175                 43                  1.0   \n",
       "2            2.754935                 39                  1.0   \n",
       "3            3.034440                 88                  1.0   \n",
       "4            2.636782                 80                  1.0   \n",
       "5            2.849910                 44                  1.0   \n",
       "6            3.033677                 46                  1.0   \n",
       "7            3.142572                 40                  1.0   \n",
       "8            2.542319                122                  1.0   \n",
       "9            2.923271                142                  1.0   \n",
       "\n",
       "   umis_coverage_25pct  umis_coverage_median  umis_coverage_75pct  \\\n",
       "0                  2.0                   3.0                  4.0   \n",
       "1                  2.0                   3.0                  5.0   \n",
       "2                  1.0                   2.0                  4.0   \n",
       "3                  1.0                   2.0                  4.0   \n",
       "4                  1.0                   2.0                  3.0   \n",
       "5                  1.0                   2.0                  4.0   \n",
       "6                  2.0                   2.0                  4.0   \n",
       "7                  2.0                   3.0                  4.0   \n",
       "8                  1.0                   2.0                  3.0   \n",
       "9                  1.0                   2.0                  4.0   \n",
       "\n",
       "   umis_coverage_95pct  umis_coverage_ge_2  umis_coverage_ge_3  \\\n",
       "0                  8.0                4211                2740   \n",
       "1                  8.0                6748                4590   \n",
       "2                  6.0                3566                2091   \n",
       "3                  7.0                5395                3303   \n",
       "4                  6.0                3485                1887   \n",
       "5                  7.0                4897                2830   \n",
       "6                  7.0                4027                2478   \n",
       "7                  7.0                5421                3367   \n",
       "8                  6.0                3313                1828   \n",
       "9                  7.0                5631                3401   \n",
       "\n",
       "   umis_coverage_ge_5  umis_coverage_ge_10  \n",
       "0                1084                  118  \n",
       "1                1941                  272  \n",
       "2                 692                   69  \n",
       "3                1265                  148  \n",
       "4                 619                   67  \n",
       "5                 972                  123  \n",
       "6                 911                   87  \n",
       "7                1247                  143  \n",
       "8                 563                   47  \n",
       "9                1252                  159  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "stats_reads = pd.read_csv('analysis/reads_parsed/stats_reads.csv')\n",
    "stats_reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To see if the probes were well-balanced and the samples were adequately covered, we can make boxplots for the number of UMIs with at least two supporting read pairs, per probe and sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5,0.98,'')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAEWCAYAAACXGLsWAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XucVXW9//HXW0RRI8CcPAooZmgk\nJXEmqJOZlxPipdCueszApkP+UqrTLxOjE6hx0uocMys7lhcsA42TxU9NJQM9U96GRELRJKCAvIxi\npHlD/Pz+WN+R5Xb2fPcwe5gR3s/HYz9m7c/6ru/6rrXX7M9e3+/aaysiMDMz68h2Pd0AMzPr/Zws\nzMwsy8nCzMyynCzMzCzLycLMzLKcLMzMLMvJwrpM0omSbupg/iGS1mzJNr1adOe+kXSvpEO6o+7e\nTNIMST/u6XZsbZwsejFJIemNFbGX/hHSG01IuqaizIEpvrCjuuolIq6MiHFbYl1dJWlHSZdI+pOk\nJyUtlnRkT7erO0TEARGxsKfbYVsHJ4tXv1bgnZJeV4pNBP7QQ+3pMkm7d2P12wOrgfcAA4AvA1dL\nGlbLwpK277aWbUE9tR1dXa8Kft/qAd7pr37PAz8HjgeQ1Af4KHBltQUkjZHUIulvkh6R9F9Vyt0i\n6YNp+l3pjOHo9PxwSYvT9CRJzWn61rT4PZKekvTRUn3/V9Kjkh6SdHIH2zRD0n2STpf0D7XuiFpE\nxN8jYkZErIqIFyPiWmAl8I/tlU/b9htJ50t6HJiR4p+QtEzSE5JulLR3aZkLJK1O+3eRpHeX5u0k\n6fK03H3A2ztqb9rnn5G0QtJjkr7R9mYpaV9Jv5b0eJp3paSBpWVXSfrnND1D0lxJP5b0N2BSJ46D\nQyStkfSltJ5Vkk4szd9R0jcl/TnV831JO1Use4akh4HLOtjH35G0XtL9kg4vzV8oaaak3wBPA2+Q\ntKekeZLWSVou6V8rqu0n6ap09vg7SQeW6ttT0v9IapW0UtJnOnoNrOBksXW4Avh4mj4CWAr8pYPy\nFwAXRMRrgX2Bq6uUuwU4JE2/B1gBHFx6fkvlAhHRNv/AiHhNRFyVnv8DxSf5wUAT8F1Jg6qs91Tg\nM8BbgQfSm8JxkvqWC0n6nqS/Vnks6WD7y3XsDuwH3NtBsbEU2747MFPSBOBLwAeABuB/gdml8ncB\no4BdgZ8AP5XUL82bTrHP96V4rSbW0MzjgEZgNDAB+ERb84GvAXsCI4ChpGRWxQRgLjCQ4sNErccB\nFK/fbhSv30TgYkn7p3nnUuzDUcAbU5mvVCy7K7A3MLlK/WOBP6Z1TAd+JmnX0vyT0rL9gT8Bc4A1\nads/BPyHpMMqtvWnbHoNfi6pb0q0/w+4J7XzcOBzko7oYNsNICL86KUPIIA3VsRmAD9O04cAa9L0\ng8D+FP9EJwKfBBa2VxdwK3AWsFtm/YcDS9L0DanO29PzW4APpOlJQHO1dqd2PgNsX4o9Cryjhn3Q\nn+LN8da0zDl13L99gV8B/91BmUnAnytivwSaSs+3o/jEu3eVOp6gSJ5QJJ3xpXmT217DDo6BcvlP\nAzdXKXsscHfp+Srgn0vHza0V5Ws9Dg4BXgB2KcWuBv6dImH9Hdi3NO+dwMrSss8D/TL7+C+ASrE7\ngZPS9ELg7NK8ocBGoH8p9jXg8tK23l7x+jwEvJsiKVW+nmcCl9XruNpaHz6z6N02UryhlfUFNrRT\n9kfAacChwDXtzC9rovgkeL+kuyQdU6XcbcB+6dP3KIozmKGSdgPGULzZ1OrxiHih9Pxp4DW5hSLi\nSWAJsJhi2/fveInapE+YP6J4IzstU3x1xfO9gQvazmKAdRRvmoNT3V9IXVTr0/wBFJ+YofgkXK7v\nTzU0t7L8nmk9u0uaI2lt6lr6cWk9tWxHrccBwBMR8fd22tEA7AwsKu2PG1K8TWtEPNtB3QBrI71z\nV9TfXtv3BNalY6NcfnB75SPiRTadhewN7Fk+C6U4S+zOcbKtgpNF7/ZnYFhFbB/af4P5EcWnzusj\n4umOKo2IByPiBOD1wHnAXEm7tFPuaWAR8FlgaUQ8D/wW+Dzwx4h4rHObUztJQyRNTf36cygG8g+M\niI+UynxfxbhIe4+q3UqSBFxC8QbxwYhoL/mWVd6aeTXwqYgYWHrsFBG/TeMTXwQ+AgyKiIHAeopk\nAsUn3KGluvbK7oxXlm/rYvyP1La3RNGV9LHSerLbUetxkAyqmNfWjscozhoPKO2LARFR/iBQy62t\nB6fXpbL+9ur4C7CrpP4V5deWnr+0z9IHgyFpudUUZz3l165/RBxVQxu3aU4WvdtVwJfTG+d2abDy\nfRT9zi8TESspxhGm5SqV9DFJDekT119T+MUqxW+h+OTdNj6xsOJ5ex4B3pBrRwftm0ExhrA/cAow\nPCLOiYg/l8tFxClRjIu09zigg1VcRNHH/76IeGYzmvh94ExJB6T2DpD04TSvP0WXTSuwvaSvAK8t\nLXt1WnaQpCHAlBrWd3oqP5QicbeNA/UHngLWSxoMnN6ZjejkcQBwlqQdUkI8BvhpWvYHwPmSXp/q\nHbwZYwCvBz6TxhU+TPH6XN9ewYhYTfGh5WuS+kl6K8VZUvm7Ff8o6QMqrr76HPAccDtF99aTacB9\nJ0l9JI2U1OGFBuZk0dudTfFP0UzR7/114MSIWNpe4YhojoiOBrbbjAfulfQUxSDn8R28ad5C8aZ0\na5Xn7ZkBzEqn+R/poFw1Pwf2jIiTI+LWiu6JLlFx1dKnKLrVHi6diZyYWfQlEXENxSfxOan7ZynQ\n9l2NGym6Yf5AcQb4LC/vQjkrxVcCN1GcEeb8guIMbzFwHcVZUVtdoynOXK4DflbrNiSdOQ4epjgG\n/0IxOH5KRNyf5p0BLAduT/vjV3S+u/AOYDjFmcpM4EMR8XgH5U+gOOv+C0W36/SI+FVp/i8orgp8\ngmJw/AMRsSEiNlIkulEUr8FjwA8pugqtA6rj/6GZ1ZmkoDizWt6DbTiE4qKKId1U/yTgkxFxUHfU\nb/XhMwszM8tysjAzsyx3Q5mZWZbPLMzMLMvJwszMsnr1HTR32223GDZsWE83w8xsq7Vo0aLHIqIh\nV65XJ4thw4bR0tLS080wM9tqSarlljPuhjIzszwnCzMzy3KyMDOzLCcLMzPLcrIwM7MsJwsz69Vm\nz57NyJEj6dOnDyNHjmT27Nn5hazuevWls2a2bZs9ezbTpk3jkksu4aCDDqK5uZmmpiYATjjhhB5u\n3balV98bqrGxMfw9C7Nt18iRI7nwwgs59NBDX4otWLCAKVOmsHRpuz/rYp0kaVFENGbLOVmYWW/V\np08fnn32Wfr23fRT9Bs2bKBfv35s3LixB1u29ag1WXjMwsx6rREjRtDc3PyyWHNzMyNGjOihFm27\nnCzMrNeaNm0aTU1NLFiwgA0bNrBgwQKampqYNi37U/NWZx7gNrNeq20Qe8qUKSxbtowRI0Ywc+ZM\nD273AI9ZmJltwzxmYWZmdeNkYWZmWU4WZmaW5WRhZmZZThZmZpZVU7KQNFDSXEn3S1om6Z2SZkha\nK2lxehxVKn+mpOWSHpB0RCk+PsWWS5raHRtkZmb1V+v3LC4AboiID0naAdgZOAI4PyK+WS4o6c3A\n8cABwJ7AryTtl2Z/F3gvsAa4S9K8iLivDtthZmbdKJssJA0ADgYmAUTE88DzkqotMgGYExHPASsl\nLQfGpHnLI2JFqndOKutkYWbWy9XSDbUP0ApcJuluST+UtEuad5qkJZIulTQoxQYDq0vLr0mxanEz\nM+vlakkW2wOjgYsi4m3A34GpwEXAvsAo4CHgP+vRIEmTJbVIamltba1HlWZm1kW1JIs1wJqIuCM9\nnwuMjohHImJjRLwI/IBNXU1rgaGl5YekWLX4y0TExRHRGBGNDQ0NndsaMzPrFtlkEREPA6sl7Z9C\nhwP3SdqjVOw4oO2XSOYBx0vaUdI+wHDgTuAuYLikfdIg+fGprJmZ9XK1Xg01BbgyvcmvAE4Gvi1p\nFBDAKuBTABFxr6SrKQauXwBOjYiNAJJOA24E+gCXRsS9ddwWMzPrJr7rrJnZNsx3nTUzs7pxsjAz\nsywnCzMzy3KyMDOzLCcLMzPLcrIwM7MsJwszM8tysjAzsywnCzMzy3KyMDOzLCcLMzPLcrIwM7Ms\nJwszM8tysjAzsywnCzMzy3KyMDOzLCcLMzPLcrIwM7MsJwszM8tysjAzs6yakoWkgZLmSrpf0jJJ\n75S0q6T5kh5MfwelspL0bUnLJS2RNLpUz8RU/kFJE7tro8zMrL5qPbO4ALghIt4EHAgsA6YCN0fE\ncODm9BzgSGB4ekwGLgKQtCswHRgLjAGmtyUYMzPr3bLJQtIA4GDgEoCIeD4i/gpMAGalYrOAY9P0\nBOCKKNwODJS0B3AEMD8i1kXEE8B8YHxdt8bMzLpFLWcW+wCtwGWS7pb0Q0m7ALtHxEOpzMPA7ml6\nMLC6tPyaFKsWfxlJkyW1SGppbW3t3NaYmVm3qCVZbA+MBi6KiLcBf2dTlxMAERFA1KNBEXFxRDRG\nRGNDQ0M9qjQzsy6qJVmsAdZExB3p+VyK5PFI6l4i/X00zV8LDC0tPyTFqsXNzKyXyyaLiHgYWC1p\n/xQ6HLgPmAe0XdE0EfhFmp4HfDxdFfUOYH3qrroRGCdpUBrYHpdiZmbWy21fY7kpwJWSdgBWACdT\nJJqrJTUBfwI+kspeDxwFLAeeTmWJiHWSzgHuSuXOjoh1ddkKMzPrViqGG3qnxsbGaGlp6elmmJlt\ntSQtiojGXDl/g9vMzLKcLMzMLMvJwszMspwszMwsy8nCzMyynCzMzCzLycLMzLKcLMzMLMvJwszM\nspwszMwsy8nCzMyynCzMzCzLycLMzLKcLMzMLMvJwszMspwszMwsq9ZfyjMzq5sDz7qJ9c9seEX8\nT+cds1n17X3Gte3GB+zUl3umj9usOu3lnCzMbItb/8wGVp179CtnnFvfX+4cNvW6uta3LXM3lJmZ\nZdWULCStkvR7SYsltaTYDElrU2yxpKNK5c+UtFzSA5KOKMXHp9hySVPrvzlmZtYdOtMNdWhEPFYR\nOz8ivlkOSHozcDxwALAn8CtJ+6XZ3wXeC6wB7pI0LyLu27ymm5nZltIdYxYTgDkR8RywUtJyYEya\ntzwiVgBImpPKOlmYmfVytY5ZBHCTpEWSJpfip0laIulSSYNSbDCwulRmTYpVi5uZWS9Xa7I4KCJG\nA0cCp0o6GLgI2BcYBTwE/Gc9GiRpsqQWSS2tra31qNLMzLqopmQREWvT30eBa4AxEfFIRGyMiBeB\nH7Cpq2ktMLS0+JAUqxavXNfFEdEYEY0NDQ2d3R4zM+sG2WQhaRdJ/dumgXHAUkl7lIodByxN0/OA\n4yXtKGkfYDhwJ3AXMFzSPpJ2oBgEn1e/TTEzs+5SywD37sA1ktrK/yQibpD0I0mjKMYzVgGfAoiI\neyVdTTFw/QJwakRsBJB0GnAj0Ae4NCLurfP2mJlZN8gmi3T10oHtxE/qYJmZwMx24tcD13eyjWZm\n1sN8uw8z2+L6j5jKW2Z1//dy+48AaOe2ItZpThZmtsU9uezcduPdcSNBqw8nCzPb4tq9iSDU/UaC\nVj++kaCZmWU5WZiZWZaThZmZZTlZmJlZlpOFmZllOVmYmVmWk4WZmWU5WZiZWZaThZmZZTlZmJlZ\nlpOFmZllOVmYmVmWk4WZmWU5WZiZWZaThZmZZTlZmJlZlpOFmZll1ZQsJK2S9HtJiyW1pNiukuZL\nejD9HZTikvRtScslLZE0ulTPxFT+QUkTu2eTzMys3jpzZnFoRIyKiMb0fCpwc0QMB25OzwGOBIan\nx2TgIiiSCzAdGAuMAaa3JRgzM+vdutINNQGYlaZnAceW4ldE4XZgoKQ9gCOA+RGxLiKeAOYD47uw\nfjMz20JqTRYB3CRpkaTJKbZ7RDyUph8Gdk/Tg4HVpWXXpFi1uJmZ9XLb11juoIhYK+n1wHxJ95dn\nRkRIino0KCWjyQB77bVXPao0M7MuqunMIiLWpr+PAtdQjDk8krqXSH8fTcXXAkNLiw9JsWrxynVd\nHBGNEdHY0NDQua0xM7NukU0WknaR1L9tGhgHLAXmAW1XNE0EfpGm5wEfT1dFvQNYn7qrbgTGSRqU\nBrbHpZiZmfVytZxZ7A40S7oHuBO4LiJuAM4F3ivpQeCf03OA64EVwHLgB8CnASJiHXAOcFd6nJ1i\nZmZVzZ49m5EjR9KnTx9GjhzJ7Nmze7pJ26TsmEVErAAObCf+OHB4O/EATq1S16XApZ1vpplti2bP\nns20adO45JJLOOigg2hubqapqQmAE044oYdbt21R8d7eOzU2NkZLS0tPN8PMesjIkSO58MILOfTQ\nQ1+KLViwgClTprB06dIebNnWQ9Ki0vfnqpdzsjCz3qpPnz48++yz9O3b96XYhg0b6NevHxs3buzB\nlm09ak0WvjeUmfVaI0aMoLm5+WWx5uZmRowY0UMt2nY5WZhZrzVt2jSamppYsGABGzZsYMGCBTQ1\nNTFt2rSebto2p9Yv5ZmZbXFtg9hTpkxh2bJljBgxgpkzZ3pwuwd4zMLMbBvmMQszM6sbJwszM8ty\nsjAzsywnCzMzy3KyMDOzLCcLMzPLcrIwM7MsJwszM8tysjAzsywnCzMzy3KyMDOzLCcLMzPLcrIw\nM7MsJwszM8uqOVlI6iPpbknXpueXS1opaXF6jEpxSfq2pOWSlkgaXapjoqQH02Ni/TfHzMy6Q2d+\n/OizwDLgtaXY6RExt6LckcDw9BgLXASMlbQrMB1oBAJYJGleRDyxuY03M7Mto6YzC0lDgKOBH9ZQ\nfAJwRRRuBwZK2gM4ApgfEetSgpgPjN/MdpuZ2RZUazfUt4AvAi9WxGemrqbzJe2YYoOB1aUya1Ks\nWvxlJE2W1CKppbW1tcbmmZlZd8omC0nHAI9GxKKKWWcCbwLeDuwKnFGPBkXExRHRGBGNDQ0N9ajS\nzMy6qJYzi3cB75e0CpgDHCbpxxHxUOpqeg64DBiTyq8FhpaWH5Ji1eJmZtbLZZNFRJwZEUMiYhhw\nPPDriPhYGodAkoBjgaVpkXnAx9NVUe8A1kfEQ8CNwDhJgyQNAsalmJmZ9XKduRqq0pWSGgABi4FT\nUvx64ChgOfA0cDJARKyTdA5wVyp3dkSs68L6zcxsC1FE9HQbqmpsbIyWlpaeboaZ2VZL0qKIaMyV\n8ze4zcwsy8nCzMyynCzMzCzLycLMzLKcLMzMLMvJwszMspwszMwsy8nCzMyynCzMzCzLycLMzLKc\nLMzMLKsrNxK0Xq64IXDn9eb7hZlZz/CZxVYsIqo+9j7j2qrzzMwqOVmYmVmWk4WZmWU5WZiZWZaT\nhZmZZTlZmJlZlpOFmZll1ZwsJPWRdLeka9PzfSTdIWm5pKsk7ZDiO6bny9P8YaU6zkzxByQdUe+N\nMTOz7tGZM4vPAstKz88Dzo+INwJPAE0p3gQ8keLnp3JIejNwPHAAMB74nqQ+XWu+mZltCTV9g1vS\nEOBoYCbweRVfDT4M+JdUZBYwA7gImJCmAeYC30nlJwBzIuI5YKWk5cAY4La6bMk27MCzbmL9Mxs6\nvdywqdd1qvyAnfpyz/RxnV6Pmb361Xq7j28BXwT6p+evA/4aES+k52uAwWl6MLAaICJekLQ+lR8M\n3F6qs7yMdcH6Zzaw6tyju309nU0uZrb1yHZDSToGeDQiFm2B9iBpsqQWSS2tra1bYpVmZpZRy5jF\nu4D3S1oFzKHofroAGCip7cxkCLA2Ta8FhgKk+QOAx8vxdpZ5SURcHBGNEdHY0NDQ6Q0yM7P6yyaL\niDgzIoZExDCKAepfR8SJwALgQ6nYROAXaXpeek6a/+so7k43Dzg+XS21DzAcuLNuW2JmZt2mK7co\nPwOYI+mrwN3AJSl+CfCjNIC9jiLBEBH3SroauA94ATg1IjZ2Yf1mZraFdCpZRMRCYGGaXkFxNVNl\nmWeBD1dZfibFFVVmZvYq4m9wm5lZlpOFmZllOVmYmVmWk4WZmWU5WZiZWZaThZmZZTlZmJlZlpOF\nmZllOVmYmVmWk4WZmWU5WZiZWZaThZmZZTlZmJlZlpOFmZllOVmYmVmWk4WZmWU5WZiZWZaThZmZ\nZTlZmJlZlpOFmZllbZ8rIKkfcCuwYyo/NyKmS7oceA+wPhWdFBGLJQm4ADgKeDrFf5fqmgh8OZX/\nakTMqufGbKv6j5jKW2ZN3QLrATi629djZr1PNlkAzwGHRcRTkvoCzZJ+meadHhFzK8ofCQxPj7HA\nRcBYSbsC04FGIIBFkuZFxBP12JBt2ZPLzmXVud3/Jj5s6nXdvg4z652y3VBReCo97Zse0cEiE4Ar\n0nK3AwMl7QEcAcyPiHUpQcwHxnet+WZmtiXUNGYhqY+kxcCjFG/4d6RZMyUtkXS+pB1TbDCwurT4\nmhSrFq9c12RJLZJaWltbO7k5ZmbWHWrphiIiNgKjJA0ErpE0EjgTeBjYAbgYOAM4u6sNioiLU300\nNjZ2dAZjJVuii2jATn27fR1m1jvVlCzaRMRfJS0AxkfEN1P4OUmXAV9Iz9cCQ0uLDUmxtcAhFfGF\nm9Fmq7A54xXDpl63RcY5zGzrkO2GktSQziiQtBPwXuD+NA5BuvrpWGBpWmQe8HEV3gGsj4iHgBuB\ncZIGSRoEjEsxMzPr5Wo5s9gDmCWpD0VyuToirpX0a0kNgIDFwCmp/PUUl80up7h09mSAiFgn6Rzg\nrlTu7IhYV79NMTOz7pJNFhGxBHhbO/HDqpQP4NQq8y4FLu1kG83MrIf5G9xmZpblZGFmZllOFmZm\nluVkYWZmWU4WZmaW1akv5dmrS/EVmA7mn9d+vLigzcxsEyeLrZjf9M2sXtwNZWZmWU4WZmaW5WRh\nZmZZThZmZpblZGFmZllOFmZmluVkYWZmWU4WZmaWpd78xS1JrcCferodW6ndgMd6uhFmneBjtnvs\nHRENuUK9OllY95HUEhGNPd0Os1r5mO1Z7oYyM7MsJwszM8tysth2XdzTDTDrJB+zPchjFmZmluUz\nCzMzy3KyMDOzLCeLGkgaKmmBpPsk3Svpsx2UnSFpraTF6XFUxfy9JD0l6QuZdW4s1bFY0rAUHyNp\noaQHJf1O0nWS3lJl3eem+EJJD0i6R9Jdkkal+M5p+fvTdp3btT31UtunpfqWpHaMTfErUzuWSrpU\nUt8Un1Aq2yLpoIr9dZOkZWn/t+2HvpLOLe2H2yQdWY/2v9rV43iVNEzSM6X49zPrXCXp96Xy/5Ti\nwyVdK+mPkhaldh2c5k2S1Fpa5ooUv1zSyhS7R9LhpfW0ewx1cX+9RtJ/l9q4UNLYjvajpG+k/5sl\nkq6RNDDFT6z4v32x9P/2ibSPlqT2T+hq27eoiPAj8wD2AEan6f7AH4A3Vyk7A/hCB3XNBX7aUZlU\n7ql2YrsDq4B/KsUOAo7taN3AQqAxTZ8MzE/TOwOHpukdgP8FjuzivnoncBuwY3q+G7Bnmj4KUHrM\nBv5Pir+GTeNnbwXur2j7e0vldk7T5wKzSuvZHfhITx8rveFRj+MVGAYs7cQ6VwG7VcT6pXW/vxQb\nCUxK05OA77RT1+XAh9L0ocCDpXntHkNd3F9zgK8B26Xn+wBHd7QfgXHA9mn6POC8dup9C/DHND0E\n+CMwoHQs79PTx0pnHv5Z1RpExEPAQ2n6SUnLgMHAfZ2pR9KxwErg75vZlNOAWRHx21LbmjtZx23A\n6WnZp4EFafp5Sb+jOKjbJely4FmgEXgt8PmIuLai2B7AYxHxXKr3pW/cRsT1pbrubFtXRDxVWn4X\nIFKZN1P8Q84vl5O0M/CvFP9sbet5BLi6E/thq1Wv47UOTgRui4h5pbYtBZZ2oo7bKNretny7x1B7\nJM0A9gXeSPGh5esR8YOKMvsCY4ETI+LFtI6VFP+nUGU/RsRNpWpuBz7UThNOoEhEAK8HngSeSvU9\n1Tb9auFuqE5K3SBvA+7ooNhp6VTzUkmD0nKvAc4AzqpxVTuVTmWvSbEDgN9llvu30nJHtDN/PPDz\nymA6jX4fcHOm/mHAGIpPXt+X1K9i/k3AUEl/kPQ9Se9pZ119gZOAG0qx4yTdD1wHfCKF9wP+Kuln\nku5Op/59KP75/xwRf8u0dZu3ucdrsk/a77dIencNq1uQjru2ddVyvH60dLye3M78asfrK46hKt4K\nHEZxxvsVSXtWzD8AWBwRGzuqJLMfPwH8sp34RynOfgDuAR4BVkq6TNL7Mu3udZwsOiG94f8P8LkO\n3qguovg0M4riU8l/pvgM4PyKT9EdeSYiRqXHcVXac4eKvvwLSuHzS8vdWIpfKWklMA34bkU921Mc\n1N+OiBWZdl0dES9GxIPACuBN5Zlp+/4RmAy0AldJmlRRx/eAWyPif0vLXRMRbwKOBc5J4e2BdwNf\nAN4OvIGi68Jq0MXj9SFgr4h4G/B54CeSXptZ5aHpuBtbpT3XpL76n5XCV5WO18tK8W9I+gPwE4pu\nnkqvOIaq+EVEPJPOcBdQfNDplI72o6RpwAvAlRXxscDT6UyKlIzGU5yB/AE4P535vGo4WdQofZL5\nH+DKiPhZtXIR8UhEbEyntD9g08E5Fvi6pFXA54AvSTqtk824FxhdWtdY4N+BATUseyLFm+0s4MKK\neRdT9At/q4Z6Kr+Y84ov6qTtXxgR0ym6zj7YNk/SdKCB4g3olZVH3Aq8QdJuwBqKT30rIuIFik+Y\no4HlwF41vHlts7p6vEbEcxHxeJpeRNHfvl8nm1F5vB5Hkex3rWHZ0yNiP4qz8UvLM3LHUIXc8Xov\ncGA6Y32FjvZj+hB0DEUXVmW9x7PprKJYceHOiPhamv9BXkWcLGogScAlwLKI+K9M2T1KT48j9c9G\nxLsjYlhEDAO+BfxHRHynk035LjBJ6UqTZOdaF04H9L8D75D0ptTer1Ikm8/VWM2HJW2X+nrfADxQ\nnilpf0nDS6FRpDsHS/okcARwQlv/cIq/Me1jJI0GdgQeB+4CBkpquyPmYRT9xU9TvB4XSNohLdcg\n6cM1bsNWrR7Ha9qffdL0G4DhFGeSnfET4F2S3l+K1Xy8Jt8BtmvrUq12DHVggqR+kl4HHEJxTL0k\nIv4ItABnlY7BYZKO7mg/ShrIxZElAAADL0lEQVQPfJFi8P7pinnbAR9h03gFkvZMx3abl/4vXi2c\nLGrzLor+0cNU5ZLYkq8rXR5HcSXHv9WrERHxMEU/6NckLZf0W4rT2pqTTkQ8Q9HVcLqkIRTdUm8G\nfpe265OZKv4M3EnRR3tKRDyb/hHaBh5fA8xScbnhklT3jDTv+xRXLd2W1vWVFP8gsFTSYoqE+NH0\nKWwjRRfUzZJ+T3EFTNsA5Zcpurnuk7QUuBbwGEahHsfrwcCS9JrMpXit13WmEelYOwY4RdIKSbdR\nvG5f7UQdkcp/MYWqHUPVLKHofrodOCci/gKQtqvNJ1Ody9OxdDnwKB3vx+9QXCE1X6+8tPhgYHVF\nl25f4JsqLrddTPF/XPWS5t7It/uwmqm4GuraiJjb020xy0ljAk9FxDd7ui1bA59ZmJlZlr9nsZkk\nfZfiNLXsgoorOnJ1vI72L1U9vG1wsSekKzwq+/9/GhGTeqA5Vgf1OF5TPXdQjCmVnRQRv+9K+7oi\nXXJb2aXzm4g4tSfas7VyN5SZmWW5G8rMzLKcLMzMLMvJwqybqLh7aWNPt8OsHpwszLqg2jd/zbY2\nThZmVaRv8t6v4jcUlkmaq+I3QFZJOk/FXXo/LGmUpNu16bcNyjfjOyl9aWuppDGp3l1U3LTvThU3\n6nt1/a6BbZOcLMw6tj/wvYgYQfEN8U+n+OMRMToi5gBXAGdExFuB3wPTS8vvHBGj0nJt9ziaBvw6\nIsZQfGv6G5J22QLbYrbZnCzMOrY6In6Tpn9M8WNTAFcBSBoADIyIW1J8FsXtHtrMhpdukPhaFbeC\nHwdMTbd9WEjxI0F7dedGmHWVv5Rn1rFqdy2t9Qes2ltewAcj4oF2ypv1Sj6zMOvYXpLemab/BXjZ\nLxNGxHrgCW36caCTgFtKRT4KoOJ3xden8jcCU0p3OX1bN7bfrC6cLMw69gBwqoqf1BxE8WNBlSZS\njDssobj19Nmlec9KupvibqlNKXYOxV1Il0i6l00/9mTWa/l2H2ZVqPgpzWsjYmQPN8Wsx/nMwszM\nsnxmYWZmWT6zMDOzLCcLMzPLcrIwM7MsJwszM8tysjAzsywnCzMzy/r/dy9xIwj08AkAAAAASUVO\nRK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAEVCAYAAAARjMm4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3Xu8VlW97/HPN8FrCJhLNwKKx1tq\nBrqXWqeLmIp4KdQuWpnodkeejeU+nVTcdRIvlGZleiqLEu+KRrnlqKl4QbNSAQXlosJWDAh0KYqa\n5vW3/5hj5WSxnjWexXrWWs+C7/v1el7MOcaY4xlzMtfzm2PMmyICMzOztryvuxtgZmb1z8HCzMyy\nHCzMzCzLwcLMzLIcLMzMLMvBwszMshwsrF0kfVnSHW3kD5e0tCvb1FN05raRNE/S8M6oe10jaYik\nkNSru9vSkzhY1Im08+7YIm28pKvT9PBU5sYWZYam9Olt1VUrEXFNRIzoiu/qKEkbSbpU0jOSXpE0\nW9Ih3d2uzhARu0fE9O5uh627HCx6libgo5I+UEobDTzZTe3pMElbd2L1vYAlwH5AX+A7wA2ShlSz\n8Lpy5Nld67GubD8rOFj0LG8C/wkcAyBpA+Bo4JpKC0jaR9JMSS9LelbSjyuUu1fSZ9P0x1KP4bA0\nf4Ck2Wn6eEn3p+n70uJzJL0q6ehSff9H0nOSlks6oY11Gi9pvqRTJf1TtRuiGhHxt4gYHxGLI+Ld\niLgZeBr459bKp3X7o6QLJb0AjE/p/yJpgaQXJd0uabvSMhdJWpK27yxJnyjlbSLp8rTcfGDvttqb\ntvk3JD0l6XlJF0h6X8rbQdLdkl5IeddI6ldadrGkA9P0eElTJF0t6WXg+HbsB8MlLZX0H+l7Fkv6\ncil/I0k/lPSXVM8vJG3SYtnTJa0ALmul/h3TvrYq1X99ldtyvKTfpHV6RdJjknaWdEbaz5ZIKvd4\np0v6vqSHUn03Sdqiwjr3TT3Q5ZKWSTo3/W1ZiYNFz3MlcFyaPhiYC/y1jfIXARdFxObADsANFcrd\nCwxP0/sBTwGfLM3f23KBiGjOHxoR74+I5j/8f6I4kh8InAj8TFL/Ct87FvgG8GHgCUlTJR0pqXe5\nkKSfS3qpwufRNta/XMfWwM7AvDaK7Uux7lsDEySNAv4DOApoAP4AXFcqPwMYBmwBXAv8RtLGKe9M\nim2+A8X/1egqmnkk0AjsBYwC/qW5+cD3gW2AXYHBpGBWwShgCtCP4mCi2v0Aiv+/LSn+/0YDEyXt\nkvLOo9iGw4AdU5nvtlh2C2A7YEwrdZ8D3AH0BwYB/6+U19a2BPg0cFVa9hHgdorfsIHA2cAvW3zX\ncRTbbwDwNnBxhfW9POXvCOwJjAD+tULZ9VdE+FMHHyCAHVukjQeuTtPDgaVpeiGwCzAZ+DLFjj29\ntbqA+4CzgC0z338A8Giavi3V+UCavxc4Kk0fD9xfqd2pna8DvUppzwEfqWIb9KH4474vLXNODbdv\nb+BO4JdtlDke+EuLtN8DJ5bm3we8BmxXoY4XKYInFEFnZClvTPP/YRv7QLn8vwF3VSh7BPBIaX4x\ncGBpv7mvRflq94PhFD+cm5XSbgD+L0XA+huwQynvo8DTpWXfBDZuo/4rgYnAoCr+z8rbcjwwrZT3\naeBVYIPSvhNAvzQ/HTivVH631LYNgCGpbC+Kg4I3gE1KZb8I3FOrfW9d+bhnUT/eofhBK+sNvNVK\n2auAk4H9gRtbyS87keJI8HFJMyQdXqHcn4Gd09H3MIo/6sGStgT2ofixqdYLEfF2af414P25hSLi\nFeBRYDbFuu/S9hLVSUM5V1H8WJycKb6kxfx2wEXNvRhgJcWP5sBU97fSENWqlN+X4qgcil5Aub5n\nqmhuy/LbpO/ZWtLkNEzyMnB16XuqWY9q9wOAFyPib620owHYFJhV2h63pfRmTRHx9zbqPo1i+z2k\n4gqu5p5TblsCPFuafh14PiLeKc3D6vtZy23ZmzW32XYpfXlpnX4JbNXGOqyXfAKqfvyF4ohnQSlt\ne1o/eX0VsAi4MiJek1Sx0ohYCHwx/WAeBUyR9IEWPwakemYBpwBzI+JNSX8Cvgn8V0Q8v/ar1jZJ\ng4BjKYYNNgSuoDii/EupzC9SmdY8ExG7V6hbwKUUR5CHRkRrwbes5WOYlwATImKN80JpTP00il7Z\nvIh4V9KLFD+GAMsphouah722zXw3rZRvHmL8XmrbHhGxUtIRwE+rXY9q94Okv6TNSnnbUgx3Pk/x\no7x7RCyr5nvXyIxYAXwVQNLHgTtVnPsaQNvbcm0MLk1vS3Hg9XyL9CUUPYstWxzgWAvuWdSP64Hv\nSBok6X3pZOWnKcadVxMRT1OcR/h2rlJJx0pqiIh3gZdS8rsVit9LceTdfH5ieov51jwL/I9cO9po\n33iKH8ddgJOAnSLinHKgAIiIk6I4L9Lap9VAkVxCMcb/6Yh4vY1ylfwCOEPS7qm9fSV9PuX1oRiy\naQJ6SfousHlp2RvSsv1TQPx6Fd93aio/mCJwN58H6kMx7LJK0kDg1PasRDv3A4CzJG2YAuLhwG/S\nsr8CLpS0Vap3oKSD29GOz6dtAcUwU6R25Lbl2jhW0m6SNqU4pzGl1BMBICKWU5xD+ZGkzdPf3g6S\n9uvgd69zHCzqx9nAn4D7Kf6IfgB8OSLmtlY4Iu6PiLZObDcbCcyT9CrFSc5j2vjRvJfij/a+CvOt\nGQ9ckbrwX6iiPS39J7BNRJwQEfdFGjSuBRVXLX2NYlhthYortl5V6eqenIi4ETgfmJyGf+YCzfdq\n3E4xDPMkxTDH31l96OOslP40xQ/SVVV85U3ALIqhuFsoekXNde0FrErpv6t2HZL27AcrKPbBv1Kc\nHD8pIh5PeadT9GofSNvjTto3XLg38GBqx1TglIh4ivy2XBtXUZy8XgFsTHEhRWuae7TzKdZ7CkVP\nx0pUw79NM+sASUHRs1rUjW0YTnFRxaBc2Xqm4ibVqyPi193dlnWFexZmZpblYGFmZlkehjIzsyz3\nLMzMLMvBwszMsur6prwtt9wyhgwZ0t3NMDNbZ82aNev5iGjIlavrYDFkyBBmzpzZ3c0wM1tnSarm\nMTQehjIzszwHCzMzy3KwMDOzLAcLMzPLcrAwM7MsBwszM8tysDAzsywHCzMzy6rrm/K6SluvJW0v\nP5jRzNZF7llQ/MDnPtudfnNV5czM1kUOFmZmluVgYWZmWQ4WZmaW5WBhZmZZDhZmZpblYGFmZlkO\nFmZmluVgYWZmWVUFC0n9JE2R9LikBZI+Kmm8pGWSZqfPoaXyZ0haJOkJSQeX0kemtEWSxnXGCpmZ\nWe1V+7iPi4DbIuJzkjYENgUOBi6MiB+WC0raDTgG2B3YBrhT0s4p+2fAQcBSYIakqRExvwbrYWZm\nnSgbLCT1BT4JHA8QEW8Cb7bxPKVRwOSIeAN4WtIiYJ+Utyginkr1Tk5lHSzMzOpcNcNQ2wNNwGWS\nHpH0a0mbpbyTJT0qaZKk/iltILCktPzSlFYp3czM6lw1waIXsBdwSUTsCfwNGAdcAuwADAOWAz+q\nRYMkjZE0U9LMpqamWlRpZmYdVE2wWAosjYgH0/wUYK+IeDYi3omId4Ff8d5Q0zJgcGn5QSmtUvpq\nImJiRDRGRGNDQ0P71sbMzDpFNlhExApgiaRdUtIBwHxJA0rFjgTmpumpwDGSNpK0PbAT8BAwA9hJ\n0vbpJPkxqayZmdW5aq+G+jpwTfqRfwo4AbhY0jAggMXA1wAiYp6kGyhOXL8NjI2IdwAknQzcDmwA\nTIqIeTVcFzMz6yRVBYuImA00tkj+ShvlJwATWkm/Fbi1PQ00M7Pu5zu4zcwsy8HCzMyyHCzMzCzL\nwcLMzLIcLMzMLMvBwszMshwszMwsy8HCzMyyHCzMzCzLwcLMzLIcLMzMLMvBwszMshwszMwsy8HC\nzMyyHCzMzCzLwcLMzLIcLMzMLKva16r2aEPPuoNVr7/V4XqGjLulQ8v33aQ3c84c0eF2mJl1tfUi\nWKx6/S0Wn3dYdzejw8HGzKy7eBjKzMyyqgoWkvpJmiLpcUkLJH1U0haSpklamP7tn8pK0sWSFkl6\nVNJepXpGp/ILJY3urJUyM7PaqrZncRFwW0R8EBgKLADGAXdFxE7AXWke4BBgp/QZA1wCIGkL4Exg\nX2Af4MzmAGNmZvUtGywk9QU+CVwKEBFvRsRLwCjgilTsCuCIND0KuDIKDwD9JA0ADgamRcTKiHgR\nmAaMrOnamJlZp6imZ7E90ARcJukRSb+WtBmwdUQsT2VWAFun6YHAktLyS1NapfTVSBojaaakmU1N\nTe1bGzMz6xTVBItewF7AJRGxJ/A33htyAiAiAohaNCgiJkZEY0Q0NjQ01KJKMzProGqCxVJgaUQ8\nmOanUASPZ9PwEunf51L+MmBwaflBKa1SupmZ1blssIiIFcASSbukpAOA+cBUoPmKptHATWl6KnBc\nuirqI8CqNFx1OzBCUv90YntESjMzszpX7U15XweukbQh8BRwAkWguUHSicAzwBdS2VuBQ4FFwGup\nLBGxUtI5wIxU7uyIWFmTtTAzs05VVbCIiNlAYytZB7RSNoCxFeqZBExqTwPNzKz7+Q5uMzPLcrAw\nM7MsBwszM8tysDAzsywHCzMzy3KwMDOzLAcLMzPLcrAwM7MsBwszM8tysDAzsywHCzMzy3KwMDOz\nLAcLMzPLcrAwM7MsBwszM8uq9uVHZmZWJUk1qad4PVB9cM/CzKzGIqLNz3an35wtU0+BAhwszMys\nCg4WZmaWVVWwkLRY0mOSZkuamdLGS1qW0mZLOrRU/gxJiyQ9IengUvrIlLZI0rjar46ZmXWG9pzg\n3j8inm+RdmFE/LCcIGk34Bhgd2Ab4E5JO6fsnwEHAUuBGZKmRsT8tWu6mZl1lc64GmoUMDki3gCe\nlrQI2CflLYqIpwAkTU5lHSzMzOpctecsArhD0ixJY0rpJ0t6VNIkSf1T2kBgSanM0pRWKd3MzOpc\ntcHi4xGxF3AIMFbSJ4FLgB2AYcBy4Ee1aJCkMZJmSprZ1NRUiyrNzKyDqhqGiohl6d/nJN0I7BMR\n9zXnS/oVcHOaXQYMLi0+KKXRRnr5uyYCEwEaGxtrcqFxn13HsccV3X8+vc+uAId1dzPMzNotGywk\nbQa8LyJeSdMjgLMlDYiI5anYkcDcND0VuFbSjylOcO8EPAQI2EnS9hRB4hjgSzVdmwpeWXAei8/r\n/h/pIeNu6e4mmJmtlWp6FlsDN6bb13sB10bEbZKukjSM4nzGYuBrABExT9INFCeu3wbGRsQ7AJJO\nBm4HNgAmRcS8Gq+PmZl1gmywSFcvDW0l/SttLDMBmNBK+q3Are1so5nVsXXxOUi2Jt/BbWYdsi4+\nB8nW5KfOWqtqdbQIPmI0Wxe4Z2GtquZI0EeMZusPBwszM8tysDAzsywHCzMzy3KwMDOzLAcLMzPL\ncrAwM7MsBwszM8tysDAzsywHCzMzy3KwMDOzLAcLMzPLcrAwM7MsBwszM8tysDAzsywHCzMzy3Kw\nMDOzLL8pz8wqGnrWHax6/a0O1zNk3C0drqPvJr2Zc+aIDtdja6eqYCFpMfAK8A7wdkQ0StoCuB4Y\nAiwGvhARL6p4H+dFwKHAa8DxEfFwqmc08J1U7bkRcUXtVsXMam3V62+x+LzDursZQG0Cjq299vQs\n9o+I50vz44C7IuI8SePS/OnAIcBO6bMvcAmwbwouZwKNQACzJE2NiBdrsB5mZl1ife1tdWQYahQw\nPE1fAUynCBajgCujePnyA5L6SRqQyk6LiJUAkqYBI4HrOtAGM7Mutb72tqo9wR3AHZJmSRqT0raO\niOVpegWwdZoeCCwpLbs0pVVKNzOzOldtz+LjEbFM0lbANEmPlzMjIiRFLRqUgtEYgG233bYWVZqZ\nWQdV1bOIiGXp3+eAG4F9gGfT8BLp3+dS8WXA4NLig1JapfSW3zUxIhojorGhoaF9a2NmZp0iGywk\nbSapT/M0MAKYC0wFRqdio4Gb0vRU4DgVPgKsSsNVtwMjJPWX1D/Vc3tN18bMzDpFNcNQWwM3FlfE\n0gu4NiJukzQDuEHSicAzwBdS+VspLptdRHHp7AkAEbFS0jnAjFTu7OaT3db16uWKDl87b9YzZINF\nRDwFDG0l/QXggFbSAxhboa5JwKT2N9NqrV6u6PC182Y9g+/gNmuH1MPusOKYyqzn8LOhzNohItr8\nbHf6zdkyDhTWEzlYmJlZ1nozDFUPY+N9N+nd3U0wM1sr60WwqMWJ3CHjbqmLE8JmZt3Bw1BmZpbl\nYGFmZlkOFmZmluVgYWZmWQ4WZmaWtV5cDWVma6fPruPY44px3d0MAPrsCuArEruLg4WZVfTKgvPq\n5pLxerhXan3mYSgzM8tysDAzsywHCzMzy3KwMDOzLAcLMzPLcrAwM7MsBwszM8tysDAzs6yqg4Wk\nDSQ9IunmNH+5pKclzU6fYSldki6WtEjSo5L2KtUxWtLC9Bld+9UxM7PO0J47uE8BFgCbl9JOjYgp\nLcodAuyUPvsClwD7StoCOBNoBAKYJWlqRLy4to03q6WhZ93Bqtff6nA9tbjTuO8mvZlz5ogO12NW\nK1UFC0mDKB7KMgH4Zqb4KODKKN5K/4CkfpIGAMOBaRGxMtU5DRgJXLeWbTerqVWvv+VHW5hVUO0w\n1E+A04B3W6RPSENNF0raKKUNBJaUyixNaZXSVyNpjKSZkmY2NTVV2TwzM+tM2Z6FpMOB5yJilqTh\npawzgBXAhsBE4HTg7I42KCImpvpobGyMjtZnrauXp4n6SaJmPUM1w1AfAz4j6VBgY2BzSVdHxLEp\n/w1JlwHfSvPLgMGl5QeltGUUQ1Hl9Olr33TriHp5mqiHW6ynqZcDLejag61ssIiIMyh6EaSexbci\n4lhJAyJiuSQBRwBz0yJTgZMlTaY4wb0qlbsd+J6k/qnciOZ6zcx6ino50IKuPdjqyPssrpHUAAiY\nDZyU0m8FDgUWAa8BJwBExEpJ5wAzUrmzm092m5lZfWtXsIiI6aSho4j4VIUyAYytkDcJmNSuFpqZ\nWbfzHdxmZpbl16qaWZvq5SKEvpv07u4mrNccLMysolqcyB0y7pa6OSFsa8/DUGZmluVgYWZmWQ4W\nZmaW5WBhZmZZDhZmZpblYGFmZlkOFmZmluVgYWZmWb4pDygenFtFufPzZYpHY5mZrVscLPAPvBXW\n1/cUmFXDwcIsWV/fU2BWDZ+zMDOzLAcLMzPLcrAwM7MsBwszM8tysDAzs6yqg4WkDSQ9IunmNL+9\npAclLZJ0vaQNU/pGaX5Ryh9SquOMlP6EpINrvTJmZtY52tOzOAVYUJo/H7gwInYEXgROTOknAi+m\n9AtTOSTtBhwD7A6MBH4uaYOONd/MzLpCVfdZSBpEcYfQBOCbKm55/hTwpVTkCmA8cAkwKk0DTAF+\nmsqPAiZHxBvA05IWAfsAf67Jmli71cO1/H6vcs9XzRMQ/PSDnq/am/J+ApwG9EnzHwBeioi30/xS\nYGCaHggsAYiItyWtSuUHAg+U6iwvY13M71a2WvGP/PohOwwl6XDguYiY1QXtQdIYSTMlzWxqauqK\nrzQzs4xqzll8DPiMpMXAZIrhp4uAfpKaeyaDgGVpehkwGCDl9wVeKKe3ssw/RMTEiGiMiMaGhoZ2\nr5CZmdVeNlhExBkRMSgihlCcoL47Ir4M3AN8LhUbDdyUpqemeVL+3VH0U6cCx6SrpbYHdgIeqtma\nmJlZp+nIgwRPByZLOhd4BLg0pV8KXJVOYK+kCDBExDxJNwDzgbeBsRHxTge+38zMuki7gkVETAem\np+mnKK5malnm78DnKyw/geKKKjMz60F8B7eZmWX5fRZmZu1UD/coQdfep+RgYWbWDuvrPUoehjIz\nsyz3LMxK1sfhBbNqOFiYJevr8IJZNTwMZWZmWQ4WZmaW5WBhZmZZDhZmZpblYGFmZlkOFmZmluVg\nYWZmWQ4WZmaW5WBhZmZZDhZmZpblYGFmZlkOFmZmluVgYWZmWQ4WZmaWlQ0WkjaW9JCkOZLmSTor\npV8u6WlJs9NnWEqXpIslLZL0qKS9SnWNlrQwfUZ33mpZR0nKfp45//CqyplZz1fN+yzeAD4VEa9K\n6g3cL+n3Ke/UiJjSovwhwE7psy9wCbCvpC2AM4FGIIBZkqZGxIu1WBGrrYjo7iaYWR3J9iyi8Gqa\n7Z0+bf2SjAKuTMs9APSTNAA4GJgWEStTgJgGjOxY883MrCtUdc5C0gaSZgPPUfzgP5iyJqShpgsl\nbZTSBgJLSosvTWmV0lt+1xhJMyXNbGpqaufqmJlZZ6gqWETEOxExDBgE7CPpQ8AZwAeBvYEtgNNr\n0aCImBgRjRHR2NDQUIsqzWrG53FsfdWuq6Ei4iXgHmBkRCxPQ01vAJcB+6Riy4DBpcUGpbRK6WY9\nRkTU5GPW01RzNVSDpH5pehPgIODxdB4CFYdJRwBz0yJTgePSVVEfAVZFxHLgdmCEpP6S+gMjUpqZ\nmdW5aq6GGgBcIWkDiuByQ0TcLOluSQ2AgNnASan8rcChwCLgNeAEgIhYKekcYEYqd3ZErKzdqpiZ\nWWfJBouIeBTYs5X0T1UoH8DYCnmTgEntbKOZmXUz38FtZmZZDhZmZpblYGFmZlkOFmZmluVgYWZm\nWdVcOmtmZu1QzV36Oj9fTz3dwOlgYWZWY/X0I18rHoYyM7MsBwszM8tysDAzsywHCzMzy3KwMDOz\nLAcLMzPLcrAwM7MsBwszM8tSPd88IqkJeKa725FsCTzf3Y2oM94ma/I2WZO3yZrqaZtsFxENuUJ1\nHSzqiaSZEdHY3e2oJ94ma/I2WZO3yZp64jbxMJSZmWU5WJiZWZaDRfUmdncD6pC3yZq8TdbkbbKm\nHrdNfM7CzMyy3LMwM7Os9TZYSJok6TlJc0tpl0t6WtIcSU9KulLSoJS3qaRbJD0uaZ6k80rLbSfp\nLkmPSprevExPthbbp4+k2aXP85J+0n1rUFuSNpD0iKSb03xvSedJWijpYUl/lnRIZj/5pqT5aT+5\nS9J23bdGHVejbXKSpMfSPnO/pN26b41WJ+kUSXNTm/89pV1f2scXS5qd0jeUdFlalzmShleo83hJ\nP62Q92qN2n25pM/Voq6y9TZYAJcDI1tJPzUihgK7AI8Ad0vaMOX9MCI+COwJfEzSIc3pwJUR8WHg\nbOD7ndryrnE57dg+EfFKRAxr/lDcH/O7rmtupzsFWFCaPwcYAHwoIvYCjgD6pLxK+8kjQGPaT6YA\nP+iSlneeWmyTayNij7TP/AD4cdc0vW2SPgR8FdgHGAocLmnHiDi6tI//lvf28a8CRMQewEHAjySt\nU7+v69TKtEdE3AesbCM/IuJCYAVwSES8FhH3pLw3gYeB5h7EbsDdafoeYFSnNbyLtHf7lPMk7Qxs\nBfyhUxvZRVLv6TDg12l+U4ofh69HxBsAEfFsRNzQ1n4SEfdExGup2gd4b//pcWq4TV4uVbsZUC8n\nUXcFHkxtfxu4FziqOVPFe1O/AFyXkv7xGxARzwEvAZXuoxicRiAWSjqzZaYKF6RezWOSjq4i/aeS\nnpB0J8XfXs2tt8GiHR4GPlhOkNQP+DRwV0qaw3s70pFAH0kf6LIWdq81tg9wDHB9rDtXT/wEOA14\nN83vCPylxQ/dGlrZT8pOBH5fy0Z2sZptE0ljJf0XRc/iG53T3HabC3xC0gdSIDwUGFzK/wTwbEQs\nTPNzgM9I6iVpe+CfW5Qv2wf4LPBh4POSWgaVo4BhFD2aA4ELJA1oI/1Iip7+bsBxwP9c+9WuzMEi\nb7U3r0vqRXE0cXFEPJWSvwXsJ+kRYD9gGfBOl7ay+7T2ZvpjeO+Iq0eTdDjwXETMaudyre0nzXnH\nUhx1XlCzhnahWm+TiPhZROwAnA58p6aNXUsRsQA4H7gDuA2Yzep/019k9X18ErAUmEkRSP9E5d+A\naRHxQkS8TjGM9fEW+R8HrouIdyLiWYpezd5tpH+ylP5X3hvlqKlenVHpOmZPVj8ynAgsjIh/nLxN\n/0FHAUh6P/DZiHipS1vZfVbbPpKGAr3a+0NSxz5GccR4KLAxsDkwHthW0uZtHEmvsZ8ASDoQ+Daw\nX/NwTQ9U021SMhm4pNaNXVsRcSlwKYCk71EEg+agdxRF76G57NvA/26el/Qn4ElJRwLNQ03/2ly8\n5Vd1RvtrzT2LCtI44DcoTtjdltLOBfoC/96i7Jalk1lnUBxlrNNa2z5JyyOuHi0izoiIQRExhKLH\ndHdEHEXxI3JR88UPkhokfT5NV9pP9gR+CXwmjWv3SDXeJjuVZg8DFlInJG2V/t2WIjhcm7IOBB6P\niKWlsptK2ixNHwS8HRHzI+LG0oUfM1PxgyRtIWkTiosA/tjiq/8AHK3iarMGip7DQ22k31dKHwDs\nX/ONwXocLCRdB/wZ2EXSUkknpqwLJM0BnqTo4u0fEW+mE3rfphgXfDhdOtd8pDAceELSk8DWwISu\nXJfO0N7tU1q0fNJvXfYdoAmYr+Ly4puBlzP7yQXA+4HfpPSp3dHwTrQ22+RkFZemzga+CYzujoZX\n8FtJ84H/D4wtjRa0Nsy6FcW6LaAYTvtKG/U+RHEl1aPAb0tBpNmNKW8OxZDSaRGxIpO+EJgPXEnx\nd1tzvoPbzMyy1tuehZmZVc/BwszMshwszMwsy8HCzMyyHCzMzCzLwcKsG6RnA/WodzDb+s3BwszM\nshwszBJJm6l478Kc9GTPoyV9V9KMND9RklLZ6ZIulDRT0gJJe0v6nYoniZ6bygxR8Q6Ha1KZKemh\ndC2/d4SKdz88LOk36ZExZnXFwcLsPSOBv0bE0Ij4EMVjTH4aEXun+U2Aw0vl34yIRuAXwE3AWOBD\nwPGlpw7vAvw8InYFXgb+rfyFkrakuPP5wPQOiJkUdzKb1RUHC7P3PEbx3J7zJX0iIlYB+0t6UNJj\nwKeA3Uvlp5aWmxcRy9PDAZ/ivcdTL4mI5mf/XM2aTxj9CMVjMP6YHnkxGujRb9CzdZOfOmuWRMST\nkvaieHfBuZLuougtNEbEEknjKZ6y2qz5qbHvlqab55v/tnJPGBXFI6u/WINVMOs07lmYJZK2AV6L\niKspHvq3V8p6Pp1HWJv3Gm/1ZcGwAAAAlUlEQVQr6aNp+kvA/S3yH6B4zeiOqQ2bqXjToFldcc/C\n7D17UDxV913gLeB/UTxCei7F62NnrEWdTwBjJU2ieCroau9riIgmSccD10naKCV/h+KpvmZ1w0+d\nNeskkoYAN6eT42Y9moehzMwsyz0LMzPLcs/CzMyyHCzMzCzLwcLMzLIcLMzMLMvBwszMshwszMws\n678BIEktkl3tZvAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stats_reads.boxplot(column='umis_coverage_ge_2', by='probe', grid=False)\n",
    "plt.title(\"UMIs with >= 2 read pairs per probe\")\n",
    "plt.suptitle(\"\")\n",
    "\n",
    "stats_reads.boxplot(column='umis_coverage_ge_2', by='sample', grid=False)\n",
    "plt.title(\"UMIs with >= 2 read pairs per sample\")\n",
    "plt.suptitle(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also visualize the relationship between read pairs and the number of UMI groups we observed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5,1,'UMIs with >= 2 read pairs vs. raw read pairs')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stats_reads.plot.scatter(x='read_pairs', y='umis_coverage_ge_2')\n",
    "plt.title(\"UMIs with >= 2 read pairs vs. raw read pairs\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### analysis/pileups/\n",
    "\n",
    "amplimap generated a pileup table from the UMI groups, telling us how many UMI groups support a given call at each basepair of the target region.\n",
    "There is one pileup table for each individual sample, as well as several summary tables that contain all of the samples aggregated together.\n",
    "\n",
    "Since we told it that our reads contained UMIs, amplimap has also grouped all raw reads sharing the same UMI into UMI groups.\n",
    "For each UMI group and each position, amplimap has determined a consensus call, which reflects the most commonly observed call among all reads in the UMI group.\n",
    "This way, biases due to DNA amplification duplicates and random sequencing errors have been minimized, resulting in highly accurate VAFs.\n",
    "\n",
    "We will use these tables to look for the mutations we wanted to validate.\n",
    "\n",
    "Based on previous work, we expected to find the following mutations:\n",
    "\n",
    "- A ``G>C`` mutation at 123276865 (GRCh37 coordinates) in samples 4C22 (at approx. 0.13%) and 4C23 (0.24%)\n",
    "- An ``A>T`` mutation at 123276893 (GRCh37 coordinates) in samples 1D29 (at approx. 2.95%) and 1D7 (0.33%).\n",
    "\n",
    "The 979-blood sample acts as a negative control which should not show any mutations. Thus, the VAF we observe in this sample gives us an idea about the error rate we can expect.\n",
    "\n",
    "All the information we need to check this is contained in the ``pileups_long_detailed.csv`` summary table, and in particular in the ``number_called_hq`` and ``count_hq_A/C/G/T`` columns. Here, the ``_hq`` denotes that the calls were filtered by base quality, with the minimum threshold being set to 30 by default."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.read_csv('analysis/pileups/pileups_long_detailed.csv')\n",
    "\n",
    "d.loc[\n",
    "    d.pos.isin([123276865, 123276893]),\n",
    "    ['sample', 'chr', 'pos', 'ref', 'number_called_hq', 'nonref_hq_count_fraction', 'count_hq_A', 'count_hq_C', 'count_hq_G', 'count_hq_T']\n",
    "].sort_values(['pos', 'sample'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first thing we see is that we thoroughly covered both sites, with around 13,000 unique UMIs observed in all five samples.\n",
    "The vast majority of these UMI groups supported the reference allele, agreeing with our expectation that any potential mutations would be observed at very low frequencies.\n",
    "The background frequency of erroneous non-reference calls also seems to be low, with ``nonref_hq_count_fraction``, which is the proportion of UMI groups with a call different from the reference base, at 0.007% and 0.05% for the blood control.\n",
    "\n",
    "Looking at samples 4C22 and 4C23 we see a non-reference call rate (``nonref_hq_count_fraction``) above the background at position 123276865 but only a low rate at position 123276893, as expected. The observed VAF for the G>C mutation at 123276865 was 15 / 12749 = 0.12% and 17 / 12924 = 0.13%. The former of these is similar to the expected level (0.13%), while the latter is a bit lower (expected 0.24%). However, a small mismatch here is not surprising given the inaccuracies involved in VAF quantification during the previous screen.\n",
    "\n",
    "Similarly, samples 1D29 and 1D7 show a non-reference call rate much higher than the background at 123276893 but low at 123276865. The VAFs for the A>T mutation are 419 / 13916 = 3.0% and 40 / 12890 = 0.31%, very close to the VAFs we expected to see."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we wanted to look at the counts for a specific sample in more detail, we could have a look at the per-sample pileup files. For example, here is all the available data for the mutation site in 1D29:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.read_csv('analysis/pileups/1D29.pileup.csv')\n",
    "d.loc[d.pos == 123276893].iloc[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is a lot of additional information here that may be helpful to understand the underlying data. For example we can see that this site was covered by two probes. Between them, we had 133357 read pairs covering this site (``raw_coverage``) which belonged to 16511 UMI groups (``umi_groups``). Of these, 2580 UMI groups didn't have at least two supporting reads (``groups_below_min``), so they were filtered out.\n",
    "\n",
    "The mean base call quality score was close to 39 for both nucleotides (``phred_A``, ``phred_T``), suggesting that it is very unlikely that they were due to a sequencing error. The high quality of the data is further confirmed by the fact that the nucleotide counts not filtered for base quality (``count_A/C/G/T``) are actually the same as the quality-filtered ones (``count_hq_A/C/G/T``), telling us that not a single UMI group had a phred score of < 20 at this site. Maybe we could consider setting an even stricter threshold to remove the few erroneous reads we observed in the blood sample."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identifying new low-frequency mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition to looking for expected mutations, we could also use amplimap to look for new ones. This could be used to identify variants present in tumour samples, but not present in matched normal tissue samples, or to identify subclonal mutations in distinct parts of a tumour. A basic approach to this could be to look for sites where in one of the samples we observed VAFs far above the background levels in one of the samples.\n",
    "\n",
    "For this part of the tutorial we will assume that we have no prior knowledge of the variants in the samples. We will analyse all of the targeted bases in exon 7 of FGFR2, a region of about 200bp.\n",
    "\n",
    "To estimate the background level, we will simply use the non-reference allele fraction of the blood sample. In the case of tumour samples this would be the matched normal sample. If we had more samples, we might be able to use something more sophisticated here, for example considering the median non-reference fraction and its variance across all samples to build a better estimate of the expected background.\n",
    "\n",
    "Let's look at the non-reference fraction in the blood sample across the entire targeted region:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.read_csv('analysis/pileups/pileups_long_detailed.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(14, 6))\n",
    "sns.lineplot(x='pos', y='nonref_hq_count_fraction', data = d[d['sample'] == '979-blood'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like the background stays below 0.1% for almost the entire targeted region, although it does go up towards the ends. This may be related to the fact that the middle region was covered by two probes, doubling our coverage and decreasing the relative error rate.\n",
    "\n",
    "Finally, we can look for sites that may show real signal, using an arbitrary threshold of 0.1% above the background level:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add nonref_hq_count_fraction from blood sample to all other samples as background_fraction \n",
    "data_with_background = d[d['sample'] != '979-blood'].merge(\n",
    "    d.loc[d['sample'] == '979-blood', ['chr', 'pos', 'nonref_hq_count_fraction']].rename(\n",
    "        columns = {'nonref_hq_count_fraction': 'background_fraction'}\n",
    "    ),\n",
    "    on = ['chr', 'pos']\n",
    ")\n",
    "data_with_background['threshold'] = (data_with_background['background_fraction'] + 1e-3)\n",
    "data_with_background['above_background'] = data_with_background['nonref_hq_count_fraction'] > data_with_background['threshold']\n",
    "\n",
    "data_with_background.loc[\n",
    "    data_with_background['above_background'],\n",
    "    ['sample', 'chr', 'pos', 'ref', 'number_called_hq', 'nonref_hq_count_fraction', 'count_hq_A', 'count_hq_C', 'count_hq_G', 'count_hq_T', 'background_fraction']\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case, we recovered the two mutations previously described, plus a third one at position 123276801 with a VAF of 0.76%/0.87% in samples 4C22 and 4C23. However, this is actually the site with the highest background level in the entire region (0.24%), so we would be extremely sceptical about this site and should probably discard it as a technical artefact.\n",
    "In fact, this position was covered by a single probe (note the coverage of this site compared to the other two positions) and is at the end of a polyA tract. Therefore, the variant may represent an artefact of slippage at the polyA tract.\n",
    "\n",
    "Nevertheless, if had a larger set of samples to work with, we could use this approach to identify potential novel mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(14, 6))\n",
    "sns.scatterplot(\n",
    "    x='pos', y='nonref_hq_count_fraction',\n",
    "    size = 'above_background', sizes = {True: 100, False: 20},\n",
    "    style = 'above_background',\n",
    "    hue = 'ref',\n",
    "    data = data_with_background\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!amplimap --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
