{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tutorial: Quantifying allele-specific expression \n",
    "\n",
    "This is the second tutorial based on *De novo and inherited loss-of-function\n",
    "variants in TLK2: identification, clinical delineation and\n",
    "genotype-phenotype evaluation of a distinct neurodevelopmental disorder*\n",
    "by MRF Reijnders, KA Miller et al and in particular Figure 4 from that paper.\n",
    "\n",
    "Here, we will discuss how amplimap was used to quantify allele-specific expression\n",
    "of two loss-of-function mutations found in patients, and thus measure their effect\n",
    "on nonsense-mediated decay (NMD).\n",
    "\n",
    "We will be using a subset of the samples, looking at a single replicate from two different mutations, both of which generate a premature stop codon:\n",
    "\n",
    "- p.Ser330\\*, which is expected to result in a truncated product leading to NMD\n",
    "- p.Arg698\\*, which is expected to escape NMD because it is located in the last exon\n",
    "\n",
    "For both of these mutations, we will look at the balance between the mutation allele and the wild-type allele.\n",
    "If no NMD occured, both the wild-type and the mutation allele should be observed in 50% of the reads. If\n",
    "one of the alleles was targeted for NMD, we should be seeing an inbalance in the read counts.\n",
    "\n",
    "## Analysis overview\n",
    "\n",
    "Starting from the raw sequencing reads, we would like to:\n",
    "\n",
    "-  Trim off primer sequences\n",
    "-  Align reads to the reference genome, taking into account gaps introduced by spliced introns\n",
    "-  Count how often each allele was observed at the mutation site "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initial setup\n",
    "\n",
    "To run this tutorial amplimap needs to be installed and configured already.\n",
    "Please see [Installation](https://amplimap.readthedocs.io/en/latest/installation.html)\n",
    "and [Configuration](https://amplimap.readthedocs.io/en/latest/configuration.html) for details.\n",
    "\n",
    "In particular, you need to have the hg19 (GRCh37) reference FASTA genome and the\n",
    "associated indices prepared for use with STAR (see [Reference genome paths](https://amplimap.readthedocs.io/en/latest/configuration.html#reference-genome-paths))."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparing the working directory\n",
    "\n",
    "\n",
    "For every experiment that we want to process, we create a new working\n",
    "directory. This will contain all the input files required, as well as\n",
    "the output generated by amplimap. This makes it easy to keep track of\n",
    "the data for each experiment, as well as to rerun analyses if required.\n",
    "\n",
    "To create a directory, we use the standard ``mkdir`` unix command and\n",
    "change into it with ``cd``:\n",
    "\n",
    "    mkdir TLK2_NMD\n",
    "    cd TLK2_NMD\n",
    "\n",
    "All further commands should now be run inside this working directory.\n",
    "\n",
    "### reads_in\n",
    "\n",
    "The first input we need to provide to amplimap is of course the\n",
    "sequencing data. These can be obtained directly from the sequencer as\n",
    "``.fastq.gz`` files and should be placed in a directory called [``reads_in``](https://amplimap.readthedocs.io/en/latest/usage.html#reads-in).\n",
    "\n",
    "[Download the sample data from this tutorial](http://userweb.molbiol.ox.ac.uk/public/koelling/amplimap/tutorial_data/TLK2_ASE.tar)\n",
    "and extract the ``reads_in`` directory into your working directory. There are many different ways of doing this but\n",
    "we recommend using ``wget`` and ``tar`` on the command line:\n",
    "\n",
    "    wget http://userweb.molbiol.ox.ac.uk/public/koelling/amplimap/tutorial_data/TLK2_ASE.tar\n",
    "    tar xf TLK2_ASE.tar\n",
    "\n",
    "You can use ``ls`` to check that the files have been extracted to the correct subdirectory:\n",
    "\n",
    "    ls reads_in\n",
    "\n",
    "This should display a list of four fastq.gz files, which represent read 1\n",
    "and read 2 of two samples:\n",
    "\n",
    "    Sample1_Ser330_L001_R1_001.fastq.gz\n",
    "    Sample1_Ser330_L001_R2_001.fastq.gz\n",
    "    Sample2_Arg698_L001_R1_001.fastq.gz\n",
    "    Sample2_Arg698_L001_R2_001.fastq.gz\n",
    "\n",
    "\n",
    "### probes.csv\n",
    "\n",
    "Next, we need to provide a [probes.csv file](https://amplimap.readthedocs.io/en/latest/usage.html#probes-csv) that describes the used\n",
    "primer sequences and the regions they are supposed to capture. This can\n",
    "be created with spreadsheet software such as Excel, as long as the file is\n",
    "saved as plain text. However, we recommend always checking the file manually\n",
    "using a plain text editor such as ``nano`` or ``vim``, to make sure it is actually in the right\n",
    "format.\n",
    "\n",
    "Create a new plain text file called ``probes.csv`` (for example using ``nano``\n",
    "or ``vim``) in your working directory and copy the following text into it:\n",
    "\n",
    "    id,first_primer_5to3,second_primer_5to3,chr,target_start,target_end,strand\n",
    "    TLK2_cDNA_1,TGCAAGACCGCTTGAGACTG,CAGCTCTGCCTGGATCTCTG,chr17,60642418,60655843,+\n",
    "    TLK2_cDNA_2,GCATGCATGTAGGGAATACCG,ACTGTTATTGGACGCCCCAG,chr17,60673966,60689893,+\n",
    "\n",
    "### snps.txt\n",
    "\n",
    "In this case we have two specific genomic positions that we want to look at and\n",
    "also know the alleles that we expect to see. Thus, we can provide a\n",
    "[snps.txt](https://amplimap.readthedocs.io/en/latest/usage.html#snps-txt) file\n",
    "and obtain allele counts specifically for these positions, rather than\n",
    "screening a whole genomic region. This both speeds up the processing\n",
    "and simplifies the downstream analysis.\n",
    "\n",
    "Create a new plain text file called ``snps.txt`` (for example using ``nano``\n",
    "or ``vim``) in your working directory and copy the following text into it:\n",
    "\n",
    "    chr17\t60650596\tTLK2_Ser330\tC\tA\n",
    "    chr17\t60689765\tTLK2_Arg698\tC\tT\n",
    "\n",
    "### config.yaml\n",
    "\n",
    "Finally, we create a config.yaml file to set some experiment-specific settings.\n",
    "We could set [a lot more options](https://amplimap.readthedocs.io/en/latest/configuration.html)\n",
    "here but in this case set a few of them. All the other options will be left at\n",
    "the as specified in the default configuration.\n",
    "\n",
    "Create a new plain text file called ``config.yaml`` (for example using ``nano``\n",
    "or ``vim``) in your working directory and copy the following text into it:\n",
    "\n",
    "    general:\n",
    "      genome_name: \"hg19\"\n",
    "    align:\n",
    "      aligner: \"star\"\n",
    "\n",
    "This tells amplimap to run using the reference genome hg19, as specified in your\n",
    "[default configuration](https://amplimap.readthedocs.io/en/latest/configuration.html#default-configuration).\n",
    "If you do not have this reference genome set up there, you can also specify the necessary paths directly\n",
    "in the ``config.yaml``, by adding the following additional lines and editing the paths to match your local setup:\n",
    "\n",
    "    paths:\n",
    "      hg19:\n",
    "        star: \"/INSERT/PATH/TO/GENOME\"\n",
    "        fasta: \"/INSERT/PATH/TO/FASTA\"\n",
    "\n",
    "For ``star`` you would provide the path to the Genome directory generated by ``STAR --runMode genomeGenerate``.\n",
    "For ``fasta`` you would provide the path to the corresponding FASTA file, which needs to have been indexed with ``samtools faidx``.\n",
    "\n",
    "Note that we also specify the **STAR aligner** instead of the normal option of BWA/bowtie2. This is because\n",
    "we are dealing with spliced cDNA here, which means that our\n",
    "reads will only contain the exonic sequence. We need to use an aligner that work with spliced data\n",
    "and create alignments with long gaps to account for introns.\n",
    "\n",
    "For a real-world analysis, we might also want to use a custom reference genome in which we have masked the target SNPs to avoid reference bias. However, for the purposes of this example, we will stick with the standard reference."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Running amplimap\n",
    "\n",
    "Now we can run amplimap. In our case, we want to obtain coverage values\n",
    "(“coverages”) and annotated variant calls (“variants”). This will also\n",
    "automatically run the other parts of the pipeline that are required,\n",
    "such as trimming the primers and aligning reads to the genome.\n",
    "First we will do a dry-run to confirm that all input files can be found:\n",
    "\n",
    "    amplimap pileups\n",
    "    \n",
    "This should output a long list of commands, ending with these lines:\n",
    "\n",
    "    Job counts:\n",
    "        count\tjobs\n",
    "        2\talign_pe\n",
    "        1\tcopy_probes\n",
    "        1\tcopy_snps\n",
    "        2\tdo_pileup_snps\n",
    "        4\tlink_reads\n",
    "        2\tparse_reads_pe\n",
    "        1\tpileup_snps_agg\n",
    "        1\tpileups\n",
    "        1\tstart_analysis\n",
    "        2\tstats_alignment\n",
    "        1\tstats_alignment_agg\n",
    "        1\tstats_reads_agg\n",
    "        1\tstats_samples_agg\n",
    "        2\ttool_version\n",
    "        22\n",
    "    amplimap dry run successful. Set --run to run!\n",
    "\n",
    "\n",
    "You can see how amplimap is planning to run 2 alignment jobs (align_pe) and 2 SNP pileup jobs (do_pileup_snps),\n",
    "corresponding to the 2 samples we are analysing.\n",
    "\n",
    "Having confirmed that everything looks as expected, we can run amplimap:\n",
    "\n",
    "    amplimap pileups --run\n",
    "\n",
    "This will take a few minutes to complete. It would be much faster if we\n",
    "ran jobs in parallel (for example using a cluster), but we are not\n",
    "doing that for the purposes of this tutorial."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysing the results\n",
    "\n",
    "amplimap has now processed our reads, aligned them to the reference genome, called germline variants, annotated them\n",
    "and produced a summary table with the variant calls.\n",
    "All of the output files have been placed into the ``analysis`` directory.\n",
    "\n",
    "Let's explore some of the output. Most analyses in amplimap produce one or more CSV file with a table of results. In this tutorial, we will use Python and pandas to process and visualize these files. However, the same thing could also be done in R or Excel.\n",
    "\n",
    "### analysis/reads_parsed/\n",
    "\n",
    "This directory contains results from the first step of the pipeline which\n",
    "identified primer arms in reads, trimmed them off and calculated some\n",
    "run statistics.\n",
    "\n",
    "For example, the ``stats_samples.csv`` file tells us about the number of reads in each sample and how many of these contained the expected primer sequences:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>files</th>\n",
       "      <th>pairs_total</th>\n",
       "      <th>pairs_unknown_arms</th>\n",
       "      <th>pairs_good_arms</th>\n",
       "      <th>pairs_r1_too_short</th>\n",
       "      <th>pairs_r2_too_short</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sample1_Ser330</td>\n",
       "      <td>1</td>\n",
       "      <td>6157</td>\n",
       "      <td>362</td>\n",
       "      <td>5795</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sample2_Arg698</td>\n",
       "      <td>1</td>\n",
       "      <td>477</td>\n",
       "      <td>26</td>\n",
       "      <td>451</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample  files  pairs_total  pairs_unknown_arms  pairs_good_arms  \\\n",
       "0  Sample1_Ser330      1         6157                 362             5795   \n",
       "1  Sample2_Arg698      1          477                  26              451   \n",
       "\n",
       "   pairs_r1_too_short  pairs_r2_too_short  \n",
       "0                   0                   0  \n",
       "1                   0                   0  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats\n",
    "%matplotlib inline\n",
    "\n",
    "pd.read_csv('analysis/reads_parsed/stats_samples.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And for a more detailed look at the number of reads observed per probe in each sample, there is ``stats_reads.csv``:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>probe</th>\n",
       "      <th>read_pairs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sample1_Ser330</td>\n",
       "      <td>TLK2_cDNA_1</td>\n",
       "      <td>5795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sample2_Arg698</td>\n",
       "      <td>TLK2_cDNA_2</td>\n",
       "      <td>451</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample        probe  read_pairs\n",
       "0  Sample1_Ser330  TLK2_cDNA_1        5795\n",
       "1  Sample2_Arg698  TLK2_cDNA_2         451"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('analysis/reads_parsed/stats_reads.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that Sample 1 had a lot more reads than Sample 2, but both of them have good coverage. As expected, Sample 1 only contained reads for the first probe targeting Ser330, while Sample 2 only contained reads for the second probe targeting Arg698."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### analysis/pileup_snps/\n",
    "\n",
    "Based on the details that we provided in ``snps.txt``, amplimap has generated pileup tables that contains the read counts for each of the alleles at each of the SNPs.\n",
    "\n",
    "Let's have a look at the detailed summary table, which tells us the count for each possible nucleotide at each position and each sample:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>snp_ref</th>\n",
       "      <th>snp_alt</th>\n",
       "      <th>number_called_hq</th>\n",
       "      <th>snp_alt_hq_count_fraction</th>\n",
       "      <th>count_hq_A</th>\n",
       "      <th>count_hq_C</th>\n",
       "      <th>count_hq_G</th>\n",
       "      <th>count_hq_T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sample1_Ser330</td>\n",
       "      <td>chr17</td>\n",
       "      <td>60650596</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>5238</td>\n",
       "      <td>0.212486</td>\n",
       "      <td>1113</td>\n",
       "      <td>3999</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sample2_Arg698</td>\n",
       "      <td>chr17</td>\n",
       "      <td>60689765</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>357</td>\n",
       "      <td>0.509804</td>\n",
       "      <td>0</td>\n",
       "      <td>175</td>\n",
       "      <td>0</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample    chr       pos snp_ref snp_alt  number_called_hq  \\\n",
       "0  Sample1_Ser330  chr17  60650596       C       A              5238   \n",
       "3  Sample2_Arg698  chr17  60689765       C       T               357   \n",
       "\n",
       "   snp_alt_hq_count_fraction  count_hq_A  count_hq_C  count_hq_G  count_hq_T  \n",
       "0                   0.212486        1113        3999           2           3  \n",
       "3                   0.509804           0         175           0         182  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d = pd.read_csv('analysis/pileups_snps/target_snps_pileups_long_detailed.csv')\n",
    "d = d.loc[d.number_called_hq > 0].sort_values(['pos', 'sample'])\n",
    "d[ ['sample', 'chr', 'pos', 'snp_ref', 'snp_alt', 'number_called_hq', 'snp_alt_hq_count_fraction', 'count_hq_A', 'count_hq_C', 'count_hq_G', 'count_hq_T'] ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To test whether the alt allele fraction differs significantly from the null expectation of 50% (ie. the fraction we would expect if there was no NMD), we can use a binomial test:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>snp_ref</th>\n",
       "      <th>snp_alt</th>\n",
       "      <th>number_called_hq</th>\n",
       "      <th>count_hq_alt</th>\n",
       "      <th>snp_alt_hq_count_fraction</th>\n",
       "      <th>p_binom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sample1_Ser330</td>\n",
       "      <td>chr17</td>\n",
       "      <td>60650596</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>5238</td>\n",
       "      <td>1113</td>\n",
       "      <td>0.212486</td>\n",
       "      <td>4.940656e-324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sample2_Arg698</td>\n",
       "      <td>chr17</td>\n",
       "      <td>60689765</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>357</td>\n",
       "      <td>182</td>\n",
       "      <td>0.509804</td>\n",
       "      <td>7.508743e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sample    chr       pos snp_ref snp_alt  number_called_hq  \\\n",
       "0  Sample1_Ser330  chr17  60650596       C       A              5238   \n",
       "3  Sample2_Arg698  chr17  60689765       C       T               357   \n",
       "\n",
       "   count_hq_alt  snp_alt_hq_count_fraction        p_binom  \n",
       "0          1113                   0.212486  4.940656e-324  \n",
       "3           182                   0.509804   7.508743e-01  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d['count_hq_alt'] = [getattr(row, 'count_hq_%s' % row.snp_alt) for row in d.itertuples()]\n",
    "d['p_binom'] = [scipy.stats.binom_test(x = row.count_hq_alt, n = row.number_called_hq, p = 0.5) for row in d.itertuples()]\n",
    "\n",
    "d[ ['sample', 'chr', 'pos', 'snp_ref', 'snp_alt', 'number_called_hq', 'count_hq_alt', 'snp_alt_hq_count_fraction', 'p_binom'] ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For Sample 1, we observed 5238 cDNA reads covering the Ser330\\* mutation site. Of these, 1113 (21%) carried the nucleotide creating the mutation, while 3999 carried the reference nucleotide. This deviation from a balanced 50%/50% allele frequency ratio is highly significant $(p < 10^{-323})$ and suggests that mRNA carrying the mutation is degraded through nonsense-mediated decay, as expected.\n",
    "\n",
    "On the other hand, we did not see a similar effect for the second mutation, which we expected to escape NMD. Here, the allele counts are roughly equal, with 51% of reads carrying the mutation $(p = 0.75)$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also visualise this with a bar plot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0,0.5,'')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcAAAAEWCAYAAADxQkdBAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBo\ndHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3Xm8FNWd/vHPoyAoIkZQoyBeooIK\nQQQ0GI1xlzhm1KiJxohLFjOa5TeZkJiJEmM0kphlYoxxjAtu4ELixqgxjqLCgAsRZRNXNrcIBhUi\nivD9/XHOlaa9S19yb/fFet6vV7+orjpV9e3q5j59TlV3KyIwMzMrmg1qXYCZmVktOADNzKyQHIBm\nZlZIDkAzMyskB6CZmRWSA9DMzArJAWjrPUknSLqnBvvdW9IzkpZJOrLa+2+PJIWkHdto2/8p6fIm\nlp8saVIr7etTkua2xrZaS/njq/RYS9pP0qK2rW795AC0ViNpnqS3cyDU37Zt5X3U5f/4HernRcT1\nEXFIa+6nQucCF0fEphFxa6Ur5T9kq0qO0QuSrpLUt6RN/eO8s2zd6ySdUzavj6TVkn7/zz6g9iwi\nfhoRX4GGXwetvK+HIqJfW2zb2g8HoLW2z+ZAqL+9VN6grf5o1cD2wKx1XHdKRGwKdAMOAt4Gpkka\nUNbuE5I+2cy2RgB/B74gqVOlBXyIngezdeIAtDZX8m79y5IWAPfl+TdLekXSG5IelNS/ZJ2NJf1S\n0vy8fJKkjYEHc5Olufe0VwNDQ5+U9Ghe79HSAJE0UdJPJE2W9JakeyT1aKL2r0p6VtLrkm6v79FK\neg74GHBHruMDwSNpO0l/kvSapCWSLi5vExGrIuK5iDgdeAA4p6zJz4Hzm6hPpAA8C1gJfLaJto09\nD8Mk/Z+kpZKekLRfyTqnSJqTj9Xzkk4r2+ZISS9LeknSqWXLDpM0O6/7oqTvNlLXfElD8vQJucb+\n+f6XJd2ap8+RdF1e7QOvg5Lt/ULS33PP+jNNHI/Bkh7P9d0s6UZJ5+Vl7w8bSvq+pPFl6/5G0kV5\nupukK/JxeFHSeZI2zMtOzq/dSms6U9JzuabZko5qrG3Zep3yPhZIelXSpfn/S0Ntt5X0x/y6fEHS\ntyrZx4eRA9Cq6dPALsCh+f5dwE7AVsBfgetL2v4CGAJ8EtgC+B6wGtg3L9889zCnlO5A0hbA/wAX\nAd2BXwH/I6l7SbMvAqfk/W4ENPaH+QDgAuDzwDbAfOAGgIjYAVjAmh7vO2XrbghMyOvUAT3r123C\nn4BPlc27BOgr6aBG1tkH6JW3fRNwUjP7gJLnQVJP0vE6j3Scvwv8UdKWue3fgMOBzUjH7NeSBufH\nODy3P5j0PJbXeAVwWkR0BQaQA7cBDwD7ldT2PGue50/n5eUaex18ApgL9CC9ebgiv0lYi6SNgFuA\nMflxjwMaC5sbgMMkdc3rbkh6TYzNy8cA7wE7ArsDhwBfKVm/opqy50ivgW7Aj4HrJG3TSNtSo4G+\nwKBcR09gVHkjSRsAdwBP5DYHAv9P0qHlbQshInzzrVVuwDxgGbA0327N8+uAAD7WxLqb5zbdSG/M\n3gZ2a6Bd/bY6lMw7GZiUp08EHilbZwpwcp6eCJxVsux04O5GaroC+HnJ/U1Jvay6ksd7UCPr7gW8\nVlpnQ/WWzR8OrCx/nLnGqXn+dcA5JetcXnKc98r1bdVITR94HoDvA9eWtfszcFIj27gV+HaevhIY\nXbKsb97+jvn+AuA0YLNmXjdfBm7P03NI4XFDvj8fGJynzwGua+Z18GzJ/U1ym482sM99gRcBlcyb\nBJyXp/cDFpUtG5GnDwaey9NbA+8AG5e0PR64v6U1NXJspgNHNPS6qT/WgIDlwA5lr78Xyh8LKYwX\nlO3jB8BVldTzYbu5B2it7ciI2Dzfyq+MXFg/IWlDSaPzcM+bpDCB9C65B9CZ9G64pbYl/dEsNZ/0\nbrfeKyXT/yAFW7PbiohlwJKybTVmO2B+RLxXQdt6PYHXG5h/ObC1pLWGN/MQ17HknnOkXtACUg+3\nKQtLprcHjs3Dn0slLSX1KrfJ+/iMpKl5CHgpcBjp+YF0fEq3VX7cj87t50t6oHSYsswDwKdyT2dD\nUk92b0l1pDdE05t5PKXef24j4h95sqHnd1vgxcgJkC1soF29saRgg3R863t/2wMdgZdLjt9/k0YX\nWloTkkZIml6yrQGsOd6N2ZIUrNNK1rs7zy+3PbBt2fP9n6QgLxwHoFVT6R+bLwJHkIbNupHe0UN6\nN7sYWAHs0Mw2GvIS6T95qd6kd/sttda2JHUhDatWsq2FQG+17EKTo4CHymdGxLuk4bCfkI5PafvN\ngEuUzqW+QgrR5oZBy//oX1vypmXziOgSEaOVzmv+kTQcvXVEbA7cWVLDy6Sgr9e7rO5HI+IIUhjc\nSgq2DxYT8Szpjcg3gQcj4k1SaHyN1ONZ3cxjWBcvAz3LhiK3a6wxcDOwn6RepONeH4ALST3AHiXH\nb7OI6N/YhhojaXvgD8A3gO75eM9k7ee8IYtJIyb9S2roFukiq3ILST3D0ue7a0Qc1tJ6PwwcgFYr\nXUl/OJaQ3r3+tH5B/oN3JfCrfMJ+Q6WLXTqRhhVXky5AacidpHNmX5TUQdIXgF1J5+NaahxwiqRB\ned8/BR6OiHkVrPsI6Y/saEldJHWWtHd5o/zY+kj6LWmo6seNbO9aUq94eMm8k0jH6eOkcz+DgL2B\n3SR9vJIHSBpS/aykQ3MtnfMFIL1I50frj/l7+eKN0o+b3AScLGlXSZsAPyp5XBspXdDSLSJWAm+S\nnrfGPED6w19/vm9i2f1yzb0OmjMFWAV8I79OjgD2bKxxRLyWa7qKFCBz8vyXgXuAX0raTNIGknaQ\n9Ol1qKkLKdhfg3QBEqkH2KT8/+UPpPOzW+V1ezZyXu8R4C2lC3s2zs/5AEl7rEO96z0HoNXKNaQh\nsxeB2cDUsuXfBWYAj5KGBX8GbJCHkM4HJuchnGGlK0XEEtJFG/9BCtfvAYdHxOKWFhgR9wJnk3pB\nL5N6pMdVuO4q0hWZO5KGJRcBXyhpspekZaRgmEjqye0RETOa2N4o0gUb5ItXDgT+KyJeKblNIw1/\nVXIxDBGxkNQT/0/SH96FwEjSsX4L+BYp6P5O6rXfXrLuXcB/kS5ueZYPXuRyIjAvD3F/HTihiVIe\nIL0perCR++V1N/k6aE7uVX+OdP5xKfAl0pukd5pYbSxpxGJs2fwRpDcLs0nHaTx5CLmFNc0GfkkK\n51dJb2wmV7j690nPwdR8vO8FPvA5xvw6Opz0ZukFUu/xctIoTOFo7SFwM7NikvQwcGlEXFXrWqw6\n3AM0s0KS9GlJH81DoCcBA0m9ZysIfxOEmRVVP9LwbhfSZw+Pyef0rCA8BGpmZoXkIVAzMyskD4G2\nIz169Ii6urpal2Fmtl6ZNm3a4oho6IP/TXIAtiN1dXU89thjtS7DzGy9Iqn8W4gq4iFQMzMrJAeg\nmZkVkgPQzMwKyQFoZmaF5AA0M7NCcgCamVkhOQDNzKyQHIBmZlZIDkAzMyskfxNMOzJn0RKGjLym\n1mWYmVXVtAtH1GS/7gGamVkhOQDNzKyQHIBmZlZIDkAzMyskB6CZmRWSA9DMzArJAWhmZoXkADQz\ns0JyAJqZWSE5AM3MrJAcgGZmVkgOQDMzKyQHoJmZFZID0MzMCskBaGZmheQANDOzQnIAmplZITkA\nzcyskByAZmZWSA5AMzMrJAegmZkVkgPQzMwKyQFoZmaF5AA0M7NCcgCamVkhOQDNzKyQHIBmZlZI\nDkAzMyskB6CZmRWSA9DMzArJAWhmZoXkADQzs0JyAJqZWSE5AM3MrJAcgGZmVkgOQDMzKyQHoJmZ\nFZID0MzMCskBaGZmheQANDOzQqo4ACX9UNIsSU9Kmi7pE21VlKSJkoY20+Ybkp6VFJJ6NNN2a0kT\nJD0habakO9ehprvz+rMkXSppwzz/JyXH5B5J2+b5knRRrvFJSYNbuk8zM2s7FQWgpL2Aw4HBETEQ\nOAhY2JaFVWByrmN+BW3PBf4SEbtFxK7AmZXuJAfZBsDnI2I3YACwJXBsbnJhRAyMiEHABGBUnv8Z\nYKd8+xrw+0r3aWZmba/SHuA2wOKIeAcgIhZHxEuSRkl6VNJMSZdJErzfg/u1pMckzZG0h6Q/SXpG\n0nm5TZ2kpyRdn9uMl7RJ+Y4lHSJpiqS/SrpZ0qa5hscjYl4L6l9UfycinizZ/sj8GJ6U9OOS2uZK\nugaYCWwXEW/mVToAGwGRt1U/H6BL/XzgCOCaSKYCm0vapsJ6zcysjXWosN09wChJTwP3AjdGxAPA\nxRFxLoCka0m9xDvyOu9GxFBJ3wZuA4YArwPPSfp1btMP+HJETJZ0JXA68Iv6neahzbOAgyJiuaTv\nA98h9eha4nfAjZK+keu/Kgf4IaQe2p6AgNsl7QssyPNPyuFVX8+fc9u7gPEl888HRgBvAPvn2T1Z\nu5e8KM97ubQwSV8j9RDp2a0jt3S9sIUPzVqi96gZtS7BzNqJinqAEbGMFGBfA14jhcnJwP6SHpY0\nAzgA6F+y2u353xnArIh4Ofcgnwe2y8sWRsTkPH0dsE/ZrocBuwKTJU0HTgK2b8Hjq6//z8DHgD8A\nOwOPS9oSOCTfHgf+mpftlFebXxp+eTuHknqTnfLjrZ//w4jYDrge+EYLa7ssIoZGxNAtumzY0odm\nZmbrqNIeIBGxCpgITMyBdxowEBgaEQslnQN0Llnlnfzv6pLp+vv1+w3WVn5fpHN3x1daZxP1vw6M\nBcZKmgDsm7d/QUT891o7leqA5Y1sZ4Wk20hDnH8pW3w9cCfwI+BF1gQ9QK88z8zM2oFKL4LpJ2mn\nklmDgLl5enE+L3fMOuy/d77ABuCLwKSy5VOBvSXtmOvoIqlvS3ci6YD684uSugI7kIY5/wycWn9e\nUVJPSVs1sP6m9efvJHUA/gV4Kt8vPS5H1M8n9YBH5ItohgFvRMRaw59mZlY7lfYANwV+K2lz4D3g\nWdJw6FLSRSKvAI+uw/7nAmfk83+zKbtSMiJey0Ot4yR1yrPPAp6W9C3ge8BHgScl3RkRX2lkP0OA\niyW9Rwr9yyPiUQBJuwBT8vU7y4AvAavK1u9COj/YKa9/P3BpXjZaUj9Sz3Y+8PU8/07gMNKx+gdw\nSsVHxczM2pwiykcdq7TjNMw4ISIG1KSAdmhgz41jwmk71rqMDzVfBGP24SNpWkQ0+dnxhvibYMzM\nrJAqvgimteXP8LVq70/SKcC3y2ZPjogzWnM/Zma2/qtZALaFiLgKuKrWdZiZWfvnIVAzMyskB6CZ\nmRWSA9DMzArJAWhmZoXkADQzs0JyAJqZWSE5AM3MrJAcgGZmVkgOQDMzKyQHoJmZFZID0MzMCskB\naGZmheQANDOzQnIAmplZITkAzcyskByAZmZWSA5AMzMrJAegmZkVkgPQzMwKyQFoZmaF5AA0M7NC\ncgCamVkhOQDNzKyQHIBmZlZIDkAzMyskB6CZmRWSA9DMzArJAWhmZoXkADQzs0JyAJqZWSE5AM3M\nrJAcgGZmVkgOQDMzKyQHoJmZFVKHWhdga2y0TX96j3qs1mWYmRWCe4BmZlZIDkAzMyskB6CZmRWS\nA9DMzArJAWhmZoXkADQzs0JyAJqZWSE5AM3MrJAcgGZmVkgOQDMzKyQHoJmZFZID0MzMCskBaGZm\nheQANDOzQnIAmplZITkAzcyskByAZmZWSA5AMzMrJAegmZkVkgPQzMwKyQFoZmaF1KHWBdgacxYt\nYcjIa2pdhtl6Y9qFI2pdgq3H3AM0M7NCcgCamVkhOQDNzKyQHIBmZlZIDkAzMyskB6CZmRWSA9DM\nzArJAWhmZoXkADQzs0JyAJqZWSE5AM3MrJAcgGZmVkgOQDMzKyQHoJmZFZID0MzMCskBaGZmheQA\nNDOzQnIAmplZITkAzcyskByAZmZWSA5AMzMrJAegmZkVkgPQzMwKyQFoZmaF5AA0M7NCcgCamVkh\nOQDNzKyQHIBmZlZIDkAzMyskB6CZmRWSA9DMzArJAWhmZoXkADQzs0JyAJqZWSE5AM3MrJAcgGZm\nVkgOQDMzKyQHoJmZFZID0MzMCskBaGZmhVRRAEr6oaRZkp6UNF3SJ9qqIEkTJQ1tps31kuZKminp\nSkkdK9jurZKm/hN1bS5pvKSnJM2RtFeev5ukKZJmSLpD0mZ5fkdJV+f5cyT9YF33bWZmra/ZAMx/\n6A8HBkfEQOAgYGFbF9aM64GdgY8DGwNfaaqxpM2BIUA3SR9rpE2HZvb5G+DuiNgZ2A2Yk+dfDpwZ\nER8HbgFG5vnHAp3y/CHAaZLqmtmHmZlVSXN/9AG2ARZHxDsAEbEYQNIo4LOkAPo/4LSICEkTgceB\nTwFdgBHAD0hhdWNEnJWD4G5gGjAYmAWMiIh/lO5Y0iHAj4FOwHPAKRGxLCLuLGnzCNCrmcfwOeAO\n4FXgOOCned0xwApgd2CypAuAscC2wBTgYFJ4rQT2BU7Ox+Bd4N287b7Ag3n6L8CfgbOBALrkYN04\nt3+zvDBJXwO+BtCzW0du6XphMw/FzOotOLfp/y+9R82oUiW2PqpkCPQeYDtJT0u6RNKn8/yLI2KP\niBhA+gN/eMk670bEUOBS4DbgDGAAcLKk7rlNP+CSiNiFFAynl+5UUg/gLOCgiBgMPAZ8p6xNR+BE\nUpg25XhgXL4dX7asF/DJiPgO8CPgvojoD4wHeuc2fYDXgKskPS7pckld8rJZwBF5+lhguzw9HlgO\nvAwsAH4REa+XFxYRl0XE0IgYukWXDZt5GGZm1lqa7QFGxDJJQ0g9uv2BGyWdCbwl6XvAJsAWpCC4\nI692e/53BjArIl4GkPQ8KSCWAgsjYnJudx3wLeAXJbseBuxK6pkBbETqlZW6BHgwIh5qrH5JWwM7\nAZNyD3WlpAERMTM3uTkiVuXpfYCj8uO+W9Lf8/wOpJ7qNyPiYUm/Ac4k9fROBS6SdHZ+3PU9wz2B\nVaTe5EeAhyTdGxHPN1armVlbWblyJYsWLWLFihW1LmWdde7cmV69etGxY7OXfVSkkiFQckBMBCZK\nmgGcBgwEhkbEQknnAJ1LVnkn/7u6ZLr+fv0+o3w3ZfcF/CUiyntsaaH0I2DLXEtTPk8KoBdykG5G\n6gX+MC9f3sz6AIuARRHxcL4/nhSARMRTwCG5pr7Av+Q2XySdM1wJ/E3SZGAo4AA0s6pbtGgRXbt2\npa6ujvy3cL0SESxZsoRFixbRp0+fVtlmJRfB9JO0U8msQcDcPL1Y0qbAMeuw7971V1KSwmJS2fKp\nwN6Sdsx1dMkBg6SvAIcCx0fE6mb2czwwPCLqIqKOdE7vuEbaTiYFZv35x48ARMQrwEJJ/XK7A4HZ\nud1W+d8NSEO2l+Y2C4AD6msn9WifaqZWM7M2sWLFCrp3775ehh+AJLp3796qPdhKzgFuClwtabak\nJ0nDkucAfwBmki76eHQd9j0XOEPSHFLQ/L50YUS8RrroZFze7xTSlZ+QQmZrYEr+WMaohnaQL7bZ\nnhSm9dt9AXijkY9y/Bg4RNJM0vm8V4C38rJvAtfnWgaRL6QBjpf0NCncXgKuyvN/B2wqaRbp+FwV\nEU82c0zMzNrM+hp+9Vq7fkWUjzy2vRxME/IFNO2GpE7Aqoh4L/dOfx8Rg6q1/4E9N44Jp+1Yrd2Z\nfej5KtA15syZwy677FLrMv5pDT0OSdPyhZct4m+CWVtv4FFJTwAXAV+tcT1mZjVz6623Iomnnkpn\nb+bNm8eAAanfMnHiRA4//PCmVq+oTS3VJAAjYl5r9/4knZKHQ0tvv2thXc9ExO4RsVv+iMe6DO2a\nmX0ojBs3jn322Ydx48bVupQ28aHpAUbEVRExqOx2Rq3rMjNbHy1btoxJkyZxxRVXcMMNNzTZdvny\n5Zx66qnsueee7L777tx2223r1KbaPjQBaGZmree2225j+PDh9O3bl+7duzNt2rRG255//vkccMAB\nPPLII9x///2MHDmS5cuXt7hNtTkAzczsA8aNG8dxx6VPjB133HFNDoPec889jB49mkGDBrHffvux\nYsUKFixY0OI21VbRB+HNzKw4Xn/9de677z5mzJiBJFatWoUkzjij4bNKEcEf//hH+vXrt9b8V199\ntdk2teQeoJmZrWX8+PGceOKJzJ8/n3nz5rFw4UL69OnDwoUN/xDQoYceym9/+1vqP1b3+OOPr1Ob\nanMAmpnZWsaNG8dRRx211ryjjz6aCy64oMH2Z599NitXrmTgwIH079+fs88+e53aVFtNPghvDfMH\n4c1alz8Iv4Y/CP9B7gGamVkhOQDNzKyQHIBmZlZIDkAzMyskB6CZmRWSA9DMzArJ3wRjZlZAQ0Ze\n06rbm3bhiFbb1kMPPcTXv/51OnbsyJQpU9h4441bbdul3AM0M7OqiwhWr17d4LLrr7+eH/zgB0yf\nPr3Nwg8cgGZmViXz5s2jX79+jBgxggEDBnDttdey1157MXjwYI499liWLVvG5Zdfzk033cTZZ5/N\nCSec0Kb1eAjUzMyq5plnnuHqq69mxx135HOf+xz33nsvXbp04Wc/+xm/+tWvGDVqFJMmTeLwww/n\nmGOOadNaHIBmZlY122+/PcOGDWPChAnMnj2bvffeG4B3332Xvfbaq6q1OADNzKxqunTpAqRzgAcf\nfHCTvzPY1nwO0MzMqm7YsGFMnjyZZ599FoDly5fz9NNPV7UG9wDNzAqoNT+2sC623HJLxowZw/HH\nH88777wDwHnnnUffvn2rVoMD0MzMqqKuro6ZM2e+f/+AAw7g0Ucf/UC7MWPGVKUeD4GamVkhOQDN\nzKyQHIBmZlZIDkAzMyskB6CZmRWSA9DMzArJH4MwMyugBed+vFW313vUjHVet66ujscee4wOHTow\nduxYTj/99FasrHHuAZqZWbuwdOlSLrnkkqrtzwFoZmZVc+SRRzJkyBD69+/PZZddttayM888k+ee\ne45BgwYxcuTINq/FQ6BmZlY1V155JVtssQVvv/02e+yxB0cfffT7y0aPHs3MmTOZPn16VWpxAJqZ\nWdVcdNFF3HLLLQAsXLiQZ555pma1OADNzKwqJk6cyL333suUKVPYZJNN2G+//VixYkXN6vE5QDMz\nq4o33niDj3zkI2yyySY89dRTTJ06da3lXbt25a233qpaPe4BtiMbbdOf3qMeq3UZZlYA/8zHFtbV\n8OHDufTSS9lll13o168fw4YNW2t59+7d2XvvvRkwYACf+cxnuPDCC9u0HgegmZlVRadOnbjrrrs+\nMH/evHnvT48dO7Zq9XgI1MzMCskBaGZmheQANDMriIiodQn/lNau3wFoZlYAnTt3ZsmSJettCEYE\nS5YsoXPnzq22TV8EY2ZWAL169WLRokW89tprtS5lnXXu3JlevXq12vYcgGZmBdCxY0f69OlT6zLa\nFQ+BmplZITkAzcyskByAZmZWSFpfrwj6MJL0FjC31nVUoAewuNZFVGB9qHN9qBFcZ2tzna2nB9Al\nIrZs6Yq+CKZ9mRsRQ2tdRHMkPeY6W8f6UCO4ztbmOltPrrFuXdb1EKiZmRWSA9DMzArJAdi+XFbr\nAirkOlvP+lAjuM7W5jpbzzrX6ItgzMyskNwDNDOzQnIAmplZITkAa0DScElzJT0r6cwGlneSdGNe\n/rCkuupXWVGd+0r6q6T3JB3TTmv8jqTZkp6U9L+Stm+ndX5d0gxJ0yVNkrRre6yzpN3RkkJSTS6R\nr+B4nizptXw8p0v6SnurMbf5fH59zpJUvZ9CX7uG5o7lr0uO49OSlrbTOntLul/S4/n/+2HNbjQi\nfKviDdgQeA74GLAR8ASwa1mb04FL8/RxwI3ttM46YCBwDXBMO61xf2CTPP1v7fhYblYy/a/A3e2x\nztyuK/AgMBUY2h7rBE4GLq52bS2scSfgceAj+f5W7bHOsvbfBK5sj3WSLob5tzy9KzCvue26B1h9\newLPRsTzEfEucANwRFmbI4Cr8/R44EBJqmKNUEGdETEvIp4EVle5tnqV1Hh/RPwj350KtN5vqVSu\nkjrfLLnbBajF1WmVvDYBfgL8DFhRzeJKVFpnLVVS41eB30XE3wEi4m9VrhFafiyPB8ZVpbK1VVJn\nAJvl6W7AS81t1AFYfT2BhSX3F+V5DbaJiPeAN4DuVamugRqyhuqstZbW+GXgrjatqGEV1SnpDEnP\nAT8HvlWl2ko1W6ekwcB2EfE/1SysTKXP+9F5KGy8pO2qU9r7KqmxL9BX0mRJUyUNr1p1a1T8fyif\nPugD3FeFuspVUuc5wJckLQLuJPVWm+QAtEKQ9CVgKHBhrWtpTET8LiJ2AL4PnFXrespJ2gD4FfAf\nta6lAncAdRExEPgLa0ZU2pMOpGHQ/Ug9qz9I2rymFTXtOGB8RKyqdSGNOB4YExG9gMOAa/NrtlEO\nwOp7ESh9N9orz2uwjaQOpO78kqpU10ANWUN11lpFNUo6CPgh8K8R8U6VaivV0mN5A3Bkm1bUsObq\n7AoMACZKmgcMA26vwYUwzR7PiFhS8lxfDgypUm31KnnOFwG3R8TKiHgBeJoUiNXUktfmcdRm+BMq\nq/PLwE0AETEF6Ez6ouzGVftkZtFvpHd9z5OGEupP5vYva3MGa18Ec1N7rLOk7RhqcxFMJcdyd9LJ\n853a+XO+U8n0Z4HH2mOdZe0nUpuLYCo5ntuUTB8FTG2HNQ4Hrs7TPUhDfN3bW5253c7APPKXp7TT\n5/wu4OQ8vQvpHGCT9Vb9gfgWkLrnT+c/zD/M884l9VAgvXO5GXgWeAT4WDutcw/Su9jlpB7qrHZY\n473Aq8D0fLu9nR7L3wCzco33NxU8tayzrG1NArDC43lBPp5P5OO5czusUaQh5dnADOC49ngs8/1z\ngNG1qK8Fx3NXYHJ+zqcDhzS3TX8VmpmZFZLPAZqZWSE5AM3MrJAcgGZmVkgOQDMzKyQHoJmZFZID\n0KzGJK0q+bb96a3x6x+Sjiz9RQlJ5+YvBGgzksblrx779zbez5ha/fqIfbh0qHUBZsbbETGosYWS\nOkT6TtiWOBKYQPqMGREx6p+or1mSPgrsERE7NtNuXR6LWZtwD9CsHcq/Z3e7pPuA/5W0af49w7/m\n3w08oqTtiNzzekLStZI+Sfqu/YgOAAAC4UlEQVRJpQtzj3KH0l6TpAPzb6bNkHSlpE55/jxJPy7Z\nx84N1NVZ0lV5+eOS9s+L7gF65v19qmydMZIulfQw8HNJXfJ+H8nbOCK3q5P0UN7/X/PjQMnF+bfg\n7gW2Ktn2aK35vcdftOJTYAXgHqBZ7W0saXqefiEijsrTg4GBEfF6/k7YoyLiTUk9gKmSbid9+8VZ\nwCcjYrGkLXL724EJETEeoP7XtCR1Jn113YER8bSka0i/k/hfeZ+LI2KwpNOB7wLlPyR7BhAR8fEc\nkPdI6ksK3AlN9GR75RpXSfopcF9EnJq//PmRHGx/Aw6OiBWSdiJ97+RQ0leZ9cuPdWtSr/ZKSd3z\nsp0jItr5F0lbO+QeoFntvR0Rg/LtqJL5f4mI1/O0gJ9KepL09W49SWFwAHBzRCwGKGnfmH6kkH06\n378a2Ldk+Z/yv9NIP3hcbh/guryvp4D5pJ/1ac7NseZXBA4BzsyhP5H01X+9gY6kX0SYQfoqwPpz\nmPsC4yJiVUS8xJqf43mD9JuEV0j6HFD/u49mFXEP0Kz9Wl4yfQKwJTAkIlbmX2Po3Ab7rP8FhVW0\n7t+H0sci4OiImFvaQNI5pO9t3Y305rzJH9yNiPck7QkcCBwDfIP0hsCsIu4Bmq0fugF/y+G3P7B9\nnn8fcGweDkTSFnn+W6SfLyo3F6iTVH+xyonAAy2o4yFSGJOHPnvnbbbEn4FvKo/LSto9z+8GvBwR\nq3NdG+b5DwJfkLShpG2A/fN6mwLdIuJO4N9JwWlWMQeg2frhemBoHh4cATwFEBGzgPOBByQ9Qfp1\nAUi/KTgyX2SyQ/1GImIFcApwc97WauDSFtRxCbBBXvdG0s/PtPQ3Fn9CGu58UtKsfL9+2yflx7Ez\na3qNtwDPkM79XQNMyfO7AhPysPAk4DstrMMKzr8GYWZmheQeoJmZFZID0MzMCskBaGZmheQANDOz\nQnIAmplZITkAzcyskByAZmZWSP8f6a0jkkc/f9gAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.barplot(\n",
    "    data=d[ ['sample', 'snp_ref_hq_count_fraction', 'snp_alt_hq_count_fraction'] ] \\\n",
    "        .rename(columns = { 'snp_ref_hq_count_fraction': 'ref',  'snp_alt_hq_count_fraction': 'alt' }) \\\n",
    "        .melt(id_vars = ['sample'], var_name = 'Allele', value_name = 'fraction'),\n",
    "    x='fraction',\n",
    "    y='sample',\n",
    "    hue='Allele',\n",
    "    orient='h',\n",
    ")\n",
    "plt.title(\"Fraction of cDNA reads with given allele\")\n",
    "plt.xlabel(\"Fraction of reads\")\n",
    "plt.ylabel(\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Appendix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#high-res version of plot\n",
    "\n",
    "plt.figure(figsize=(5, 3), dpi=400)\n",
    "sns.barplot(\n",
    "    data=d[ ['sample', 'snp_ref_hq_count_fraction', 'snp_alt_hq_count_fraction'] ] \\\n",
    "        .rename(columns = { 'snp_ref_hq_count_fraction': 'ref',  'snp_alt_hq_count_fraction': 'alt' }) \\\n",
    "        .melt(id_vars = ['sample'], var_name = 'Allele', value_name = 'fraction'),\n",
    "    x='fraction',\n",
    "    y='sample',\n",
    "    hue='Allele',\n",
    "    orient='h',\n",
    ")\n",
    "plt.title(\"Fraction of cDNA reads with given allele\")\n",
    "plt.xlabel(\"Fraction of reads\")\n",
    "plt.ylabel(\"\")\n",
    "sns.despine(bottom=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!amplimap --version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
